{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535174b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from rich import inspect, print as rprint\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "\n",
    "from ALLCools.clustering import *\n",
    "from ALLCools.integration.seurat_class import SeuratIntegration\n",
    "from spida.P.setup_adata import _calc_embeddings\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from ALLCools.plot import *\n",
    "from spida.pl import plot_categorical, plot_continuous\n",
    "# Move Plot ALLCools functions outside of the cli\n",
    "plt.rcParams['axes.facecolor'] = 'white'\n",
    "\n",
    "from PyComplexHeatmap import *\n",
    "\n",
    "from datetime import datetime \n",
    "current_datetime = datetime.now().strftime(\"%Y-%m-%d_%H:%M\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ab576a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#parameters\n",
    "EXP = None\n",
    "REF_EXP = None\n",
    "\n",
    "# ref_adata_path=\"/home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}.h5ad\"\n",
    "# spatial_adata_path=\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}.h5ad\"\n",
    "\n",
    "\n",
    "# Paths\n",
    "integrator_path=\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_joint.h5ad\"\n",
    "anndata_store_path = \"/home/x-aklein2/projects/aklein/BICAN/data/aggregated\"\n",
    "annotations_store_path = \"/home/x-aklein2/projects/aklein/BICAN/data/annotated\"\n",
    "image_path = \"/home/x-aklein2/projects/aklein/BICAN/BG/images/annotations\"\n",
    "cell_mapping_out_path = \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_cell_mappings.tsv\"\n",
    "\n",
    "mc_subset_col=\"Subclass\"\n",
    "mc_subset_value=None\n",
    "spatial_subset_col=\"c2c_allcools_label_Subclass\"\n",
    "spatial_subset_value=None\n",
    "\n",
    "# umap_coord_base=\"base_round1_umap\"\n",
    "# tsne_coord_base=\"base_round1_tsne\"\n",
    "\n",
    "## Spatial related params\n",
    "spatial_adata_path=\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}.h5ad\"\n",
    "spatial_ann_path = None # .tsv file of annotations \n",
    "spatial_ann_cols=[\"Neighborhood\", \"Subclass\"]\n",
    "spatial_cell_type = \"all_round1_leiden\"\n",
    "spatial_downsample=None\n",
    "normalize_spatial=False\n",
    "spatial_query=None\n",
    "spatial_std_cutoff=None\n",
    "spatial_tsne_key='base_round1_tsne'\n",
    "spatial_umap_key='base_round1_umap'\n",
    "spatial_batch_key=\"MERSCOPE\"\n",
    "harmony_batch_key=[\"replicate\", \"donor\"]\n",
    "added_joint_emb_key =  \"subclass_\"\n",
    "\n",
    "\n",
    "## mC related Params\n",
    "mc_adata_path = \"/home/x-aklein2/projects/aklein/BICAN/data/reference/mC/BG.gene-CHN.h5ad\"\n",
    "mc_ann_path = \"/home/x-aklein2/projects/aklein/BICAN/data/reference/mC/annotations.tsv\"\n",
    "mc_ann_cols = [\"Class\", \"CellClass\", \"Subclass\", \"Group\", \"Region\"]\n",
    "mc_query = \" CellClass != 'Nonneuron' \"\n",
    "mc_cell_type = \"Group\"\n",
    "mc_std_cutoff = 0.01\n",
    "mc_downsample = 2000\n",
    "mc_palette = None\n",
    "level=None\n",
    "key_to_transfer=\"Group\"\n",
    "normalize_mc_per_cell=True\n",
    "mc_batch_key=\"mC\"\n",
    "\n",
    "# General Factors: \n",
    "min_cell=20\n",
    "topn=200\n",
    "n_train_cell=100000\n",
    "chunk_size = 50000\n",
    "resolution = 4\n",
    "cpu=16\n",
    "ref_palette=None\n",
    "spatial_palette=None\n",
    "use_genes = None\n",
    "k_weight_transfer = 20\n",
    "\n",
    "outdir=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f681a43e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_adata_path = mc_adata_path.format(REF_EXP=REF_EXP)\n",
    "spatial_adata_path = spatial_adata_path.format(EXP=EXP)\n",
    "integrator_path = integrator_path.format(EXP=EXP)\n",
    "cell_mapping_out_path = cell_mapping_out_path.format(EXP=EXP)\n",
    "Path(integrator_path).parent.mkdir(parents=True, exist_ok=True)\n",
    "Path(cell_mapping_out_path).parent.mkdir(parents=True, exist_ok=True)\n",
    "image_path = Path(image_path) / EXP\n",
    "image_path.mkdir(parents=True, exist_ok=True)\n",
    "ref_cell_type = mc_cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ec1a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LABEL_COL = qry_subset_col\n",
    "# TRANSFER_LEVEL = mc_cell_type\n",
    "# added_joint_emb_key = \"subclass_\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7adc369",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b04b99f5",
   "metadata": {},
   "source": [
    "### Spatial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c205ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_adata = ad.read_h5ad(spatial_adata_path)\n",
    "spatial_adata = spatial_adata[spatial_adata.obs[spatial_subset_col] == spatial_subset_value].copy()\n",
    "\n",
    "# To dense\n",
    "if issparse(spatial_adata.X):\n",
    "    spatial_adata.X = spatial_adata.X.toarray()\n",
    "# std cutoff\n",
    "if spatial_std_cutoff is not None:\n",
    "    std_filter = spatial_adata.X.std(axis=0) > spatial_std_cutoff\n",
    "    spatial_adata._inplace_subset_var(std_filter)\n",
    "\n",
    "spatial_adata.strings_to_categoricals()\n",
    "spatial_genes = spatial_adata.var_names.tolist()\n",
    "\n",
    "adata_cp = spatial_adata.copy()\n",
    "sc.experimental.pp.highly_variable_genes(adata_cp, n_top_genes=200, subset=True, layer='counts')\n",
    "# TODO: \n",
    "_calc_embeddings(\n",
    "    adata_cp,\n",
    "    layer=\"volume_norm\",\n",
    "    key_added=added_joint_emb_key,\n",
    "    leiden_res=2.5,\n",
    "    min_dist=0.25,\n",
    "    knn=50,\n",
    "    run_harmony=True, \n",
    "    batch_key=harmony_batch_key,\n",
    "    harmony_nclust=20,\n",
    "    max_iter_harmony=20,\n",
    ")\n",
    "# Copy Over results\n",
    "for col in adata_cp.obs.columns: \n",
    "    if col not in spatial_adata.obs: \n",
    "        spatial_adata.obs[col] = adata_cp.obs[col].copy()\n",
    "for key, value in adata_cp.obsm.items(): \n",
    "    if key not in spatial_adata.obsm: \n",
    "        spatial_adata.obsm[key] = value.copy()\n",
    "\n",
    "spatial_tsne_key=f'{added_joint_emb_key}tsne'\n",
    "spatial_umap_key=f'{added_joint_emb_key}umap'\n",
    "spatial_cell_type=f\"{added_joint_emb_key}leiden\"\n",
    "\n",
    "sc.tl.rank_genes_groups(spatial_adata,groupby=spatial_cell_type,method=\"wilcoxon\",use_raw=False)\n",
    "\n",
    "spatial_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3fc143",
   "metadata": {},
   "source": [
    "### Methylation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73d96db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in ref\n",
    "ref_adata = ad.read_h5ad(ref_adata_path, backed='r')\n",
    "# Add the annotations from file \n",
    "keep_cells=ref_adata.obs.index.tolist()\n",
    "if not mc_ann_path is None:\n",
    "    if mc_ann_path.endswith('.h5ad'):\n",
    "        obs=ad.read_h5ad(os.path.expanduser(mc_ann_path)).obs.copy()\n",
    "    elif mc_ann_path.endswith('.csv') or mc_ann_path.endswith('.csv.gz'):\n",
    "        obs=pd.read_csv(os.path.expanduser(mc_ann_path),index_col=0)\n",
    "    else:\n",
    "        obs=pd.read_csv(os.path.expanduser(mc_ann_path),index_col=0,sep='\\t')\n",
    "    # adata.obs=obs.reindex(index=adata.obs_names).copy()\n",
    "    keep_cells=list(set(keep_cells) & set(obs.index.tolist()))\n",
    "    if mc_ann_cols is None:\n",
    "        mc_ann_cols=[]\n",
    "    for col in obs.columns.tolist():\n",
    "        if col not in mc_ann_cols+[mc_cell_type]:\n",
    "            continue\n",
    "        ref_adata.obs[col]=ref_adata.obs_names.map(obs[col].to_dict())\n",
    "\n",
    "if not level is None and mc_cell_type=='cluster_id' and mc_cell_type in ref_adata.obs.columns.tolist():\n",
    "    # too many clusters for cluster_id, use the first two level (such as  c5+c20+c0 -> c5+c20)\n",
    "    ref_adata.obs[mc_cell_type]=ref_adata.obs['cluster_id'].apply(lambda x:\"+\".join(x.split('+')[:level]) if not pd.isna(x) else x)\n",
    "\n",
    "# Downsample\n",
    "ref_adata.obs[mc_cell_type]=ref_adata.obs[mc_cell_type].astype('category')\n",
    "vc=ref_adata.obs[mc_cell_type].value_counts()\n",
    "keep_cts=vc[vc>=min_cell].index.tolist()\n",
    "cells=ref_adata.obs.loc[ref_adata.obs[mc_cell_type].isin(keep_cts)].index.tolist()\n",
    "keep_cells=list(set(keep_cells) & set(cells))\n",
    "if mc_downsample and ref_adata.n_obs > mc_downsample:\n",
    "    if not mc_query is None:\n",
    "        use_cells=ref_adata.obs.loc[ref_adata.obs[mc_cell_type].notna()].query(mc_query).groupby(mc_cell_type).apply(lambda x:x.sample(mc_downsample).index.tolist() if x.shape[0] > mc_downsample else x.index.tolist()).sum()\n",
    "    else:\n",
    "        use_cells=ref_adata.obs.loc[ref_adata.obs[mc_cell_type].notna()].groupby(mc_cell_type).apply(lambda x:x.sample(mc_downsample).index.tolist() if x.shape[0] > mc_downsample else x.index.tolist()).sum()\n",
    "    keep_cells=list(set(keep_cells) & set(use_cells))\n",
    "elif not mc_query is None:\n",
    "    use_cells=ref_adata.obs.loc[ref_adata.obs[mc_cell_type].notna()].query(mc_query).index.tolist()\n",
    "    keep_cells=list(set(keep_cells) & set(use_cells))\n",
    "ref_adata=ref_adata[keep_cells,:].to_memory()\n",
    "# final setup\n",
    "ref_adata.obs[mc_cell_type]=ref_adata.obs[mc_cell_type].astype('category')\n",
    "ref_adata.strings_to_categoricals()\n",
    "\n",
    "# Get the genes from the spatial data\n",
    "common_genes = list(set(spatial_genes) & set(ref_adata.var_names.tolist()))\n",
    "ref_adata = ref_adata[:,common_genes].copy()\n",
    "\n",
    "#Subset the mC data further: \n",
    "ref_adata = ref_adata[ref_adata.obs[mc_subset_col] == mc_subset_value].copy() if mc_subset_value is not None else ref_adata\n",
    "\n",
    "# Get marker genes\n",
    "sc.tl.rank_genes_groups(ref_adata,groupby=ref_cell_type,method=\"wilcoxon\")\n",
    "\n",
    "# If Normalize mc\n",
    "if normalize_mc_per_cell:  # divide frac by prior mean (determined by alpha and beta) for each cell\n",
    "    cols = ref_adata.obs.columns.tolist()\n",
    "    if 'prior_mean' in cols:\n",
    "        print(\"Normalizing cell level fraction by alpha and beta (prior_mean)\")\n",
    "        ref_adata.X = ref_adata.X / ref_adata.obs.prior_mean.values[:, None]\n",
    "ref_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89d1c68b",
   "metadata": {},
   "outputs": [],
   "source": [
    "vc = ref_adata.obs[mc_cell_type].value_counts()\n",
    "print(vc)\n",
    "\n",
    "if len(vc[vc > 20].index) ==  1: \n",
    "    print(f\"Only one group with > 20 cells found in {mc_subset_col}, skipping\")\n",
    "    annot = vc.index[0]\n",
    "\n",
    "    key = mc_cell_type\n",
    "    \n",
    "    spatial_adata.obs[f'infer_{key}'] = annot\n",
    "    spatial_adata.obs[f'infer_{key}_prob'] = 1.0\n",
    "    spatial_adata.obs[f'infer_{key}_c2c'] = annot\n",
    "    \n",
    "    Path(integrator_path).parent.mkdir(parents=True, exist_ok=True)\n",
    "    spatial_adata.write_h5ad(integrator_path)\n",
    "    sys.exit(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fc2ebb4",
   "metadata": {},
   "source": [
    "## View"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9834fb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spida.pl import plot_categorical\n",
    "plt.rcParams['axes.facecolor'] = 'white'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0638c923",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_colors=None\n",
    "if spatial_tsne_key is not None:\n",
    "    with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):\n",
    "        plt.figure()\n",
    "        plot_categorical(spatial_adata, cluster_col=spatial_cell_type, coord_base=spatial_tsne_key, show=False)\n",
    "        # plt.savefig(os.path.join(\"figures\",\"MajorType.tsne.summary.pdf\"),bbox_inches='tight',dpi=300)\n",
    "        plt.show()\n",
    "if spatial_umap_key is not None:\n",
    "    with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):\n",
    "        plt.figure()\n",
    "        plot_categorical(spatial_adata, cluster_col=spatial_cell_type, coord_base=spatial_umap_key, show=False)\n",
    "        # plt.savefig(os.path.join(\"figures\",\"MajorType.umap.summary.pdf\"),bbox_inches='tight',dpi=300)\n",
    "        plt.show()\n",
    "    spatial_colors={cluster:color for cluster,color in zip(spatial_adata.obs[spatial_cell_type].cat.categories.tolist(),spatial_adata.uns[f'{spatial_cell_type}_colors'])}\n",
    "spatial_adata.obs[spatial_cell_type].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3e3135e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):\n",
    "    plt.figure()\n",
    "    sc.pl.umap(ref_adata,color=[ref_cell_type],cmap='jet',\n",
    "           ncols=2,wspace=0.25,show=False,vmin='p5',vmax='p95')\n",
    "    plt.show()\n",
    "ref_colors={cluster:color for cluster,color in zip(ref_adata.obs[ref_cell_type].cat.categories.tolist(),ref_adata.uns[f'{ref_cell_type}_colors'])}\n",
    "ref_adata.obs[ref_cell_type].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5346337",
   "metadata": {},
   "source": [
    "## Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bbdb637",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ref_adata.var_names.isin(spatial_adata.var_names).sum(),spatial_adata.var_names.isin(ref_adata.var_names).sum())\n",
    "ref_adata._inplace_subset_var(ref_adata.var_names.isin(spatial_adata.var_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a5f99eb",
   "metadata": {},
   "source": [
    "### mC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3907cf46",
   "metadata": {},
   "outputs": [],
   "source": [
    "if use_genes is None:\n",
    "    # use top 200 DEGs and DMGs for each cell type group and use the union of two gene sets as features for integration\n",
    "    markers = sc.get.rank_genes_groups_df(ref_adata, group=ref_adata.obs[ref_cell_type].unique())\n",
    "    markers=markers.loc[~ markers.names.isna()]\n",
    "    markers=markers.loc[(~ markers.logfoldchanges.isna()) & (markers.scores < 0) & (markers.pvals < 0.05)]\n",
    "    markers.sort_values('logfoldchanges',ascending=True,inplace=True)\n",
    "    ref_features=np.unique(markers.groupby('group').apply(lambda x:x.head(topn).names.tolist()).sum())\n",
    "    print(len(ref_features),ref_features[:10])\n",
    "    # markers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b8f582f",
   "metadata": {},
   "source": [
    "### Spatial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f39004cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if use_genes is None:\n",
    "    markers = sc.get.rank_genes_groups_df(spatial_adata, group=spatial_adata.obs[spatial_cell_type].unique())\n",
    "    markers=markers.loc[~ markers.names.isna()]\n",
    "    markers=markers.loc[(~ markers.logfoldchanges.isna()) & (markers.scores > 0) & (markers.pvals < 0.05)]\n",
    "    markers.sort_values('logfoldchanges',ascending=False,inplace=True)\n",
    "    spatial_features=np.unique(markers.groupby('group').apply(lambda x:x.head(topn).names.tolist()).sum())\n",
    "    print(len(spatial_features),spatial_features[:10])\n",
    "    # markers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d0fb9c2",
   "metadata": {},
   "source": [
    "### combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2ce92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if use_genes is None:\n",
    "    selected_features=np.unique(np.concatenate((ref_features, spatial_features)))\n",
    "else:\n",
    "    if isinstance(use_genes,list):\n",
    "        selected_features=use_genes\n",
    "    else:\n",
    "        print(use_genes)\n",
    "        selected_features=pd.read_csv(os.path.expanduser(use_genes),index_col=0,sep='\\t',header=None).index.tolist()\n",
    "print(len(selected_features),selected_features[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80cda842",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_adata._inplace_subset_var(ref_adata.var_names.isin(selected_features))\n",
    "spatial_adata._inplace_subset_var(spatial_adata.var_names.isin(selected_features))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceb2f81e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_adata, spatial_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66df9d99",
   "metadata": {},
   "source": [
    "## Filter + Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b84440b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not mc_std_cutoff is None:\n",
    "    std_filter = ref_adata.X.std(axis=0) > mc_std_cutoff\n",
    "    ref_adata._inplace_subset_var(std_filter)\n",
    "ref_adata\n",
    "\n",
    "# Normalize mc adata\n",
    "# log mC fraction and scale features\n",
    "log_scale(ref_adata, with_mean=True) ## after log: 0-0.69, after scale: z score (columns)\n",
    "# reverse mC fraction so its positively corr with RNA\n",
    "ref_adata.X *= -1\n",
    "ref_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a19fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_vars=list(set(ref_adata.var_names) & set(spatial_adata.var_names))\n",
    "ref_adata=ref_adata[:,use_vars]\n",
    "spatial_adata=spatial_adata[:,use_vars]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ccbbc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335fc966",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d998e4",
   "metadata": {},
   "source": [
    "## Merge + PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6516024",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_adata=ref_adata\n",
    "query_adata=spatial_adata\n",
    "batch_categories=[mc_batch_key, spatial_batch_key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01894c9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ref_adata.concatenate(\n",
    "    query_adata, batch_categories=batch_categories, batch_key=\"Modality\", index_unique=None\n",
    ")\n",
    "print(adata.obs.Modality.value_counts())\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb18ebb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "# select ref cells to fit the model\n",
    "train_cell = np.zeros(ref_adata.shape[0], dtype=bool)\n",
    "if ref_adata.shape[0] > n_train_cell:\n",
    "    train_cell[\n",
    "        np.random.choice(np.arange(ref_adata.shape[0]), n_train_cell, False)\n",
    "    ] = True\n",
    "else:\n",
    "    train_cell[:] = True\n",
    "\n",
    "ref_adata.obs[\"Train\"] = train_cell.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "083ad618",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run PCA on Merged Adata\n",
    "n_feature = ref_adata.shape[1]\n",
    "if n_feature >= 100:\n",
    "    model = TruncatedSVD(n_components=100, random_state=0, algorithm='randomized')\n",
    "else:\n",
    "    model = TruncatedSVD(n_components=n_feature - 1, random_state=0, algorithm='randomized')\n",
    "\n",
    "# use selected train cells to fit\n",
    "model.fit(ref_adata.X[ref_adata.obs[\"Train\"].values])\n",
    "sel_dim = model.singular_values_ != 0\n",
    "print(sel_dim.sum(), 'non-zero singular values')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bda96e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "ax.plot(model.explained_variance_ratio_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6396559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform all other data\n",
    "chunks = []\n",
    "for chunk_start in range(0, adata.shape[0], chunk_size):\n",
    "    chunks.append(\n",
    "        model.transform(adata.X[chunk_start : (chunk_start + chunk_size)])\n",
    "    )\n",
    "\n",
    "# add NNZ PCs to adata\n",
    "adata.obsm[\"X_pca\"] = np.concatenate(chunks, axis=0)[:, sel_dim]\n",
    "\n",
    "# remove low variance PCs\n",
    "n_pcs = significant_pc_test(adata, p_cutoff=0.05, obsm=\"X_pca\")\n",
    "\n",
    "# scale PC by singular values\n",
    "adata.obsm[\"X_pca\"] /= model.singular_values_[sel_dim][:n_pcs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "358210d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in adata.obs.select_dtypes(['category', 'object']).columns:\n",
    "    adata.obs[col] = adata.obs[col].astype(str)\n",
    "for i, m in enumerate(batch_categories):\n",
    "    adata1 = adata[adata.obs[\"Modality\"] == m]\n",
    "    cols=adata1.obs.columns.tolist()\n",
    "    for col in cols:\n",
    "        if adata1.obs[col].notna().sum() == 0:\n",
    "            adata1.obs.drop(col,axis=1,inplace=True) #otherwise, would get error:Can't implicitly convert non-string objects to strings for all nan columns\n",
    "    adata1.strings_to_categoricals()\n",
    "    if i == 0: \n",
    "        ref_adata = adata1.copy()\n",
    "    elif i == 1:\n",
    "        spatial_adata = adata1.copy()\n",
    "    # adata1.write_h5ad(f\"{m.lower()}_pca.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e4cdaa9",
   "metadata": {},
   "source": [
    "# Seurat Integration "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977497ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_adata = spatial_adata\n",
    "adata_list = [ref_adata, query_adata]\n",
    "for adata in adata_list:\n",
    "    print(adata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "109599b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells = sum([a.shape[0] for a in adata_list])\n",
    "features = adata_list[0].shape[1]\n",
    "\n",
    "adata_merge = ad.AnnData(\n",
    "    X=csr_matrix((cells, features), dtype=np.float32),\n",
    "    obs=pd.concat([a.obs for a in adata_list]),\n",
    "    var=adata_list[0].var,\n",
    ")\n",
    "print(adata_list[0].obsm[\"X_pca\"].shape,adata_list[1].obsm[\"X_pca\"].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8302f31c",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_pc = adata_list[0].obsm[\"X_pca\"].shape[1]\n",
    "if n_pc < 10:\n",
    "    n_cca_components = n_pc\n",
    "else:\n",
    "    n_cca_components = max(n_pc - 10, 10)\n",
    "\n",
    "print(\"CCA Components\", n_cca_components)\n",
    "min_sample = adata_merge.obs[\"Modality\"].value_counts().min()\n",
    "print(\"Smaller Sample Size\", min_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19648f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "integrator = SeuratIntegration()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cac0fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take ~2.5-3h for 300K mC + 4M 10X RNA\n",
    "anchor = integrator.find_anchor(\n",
    "    adata_list,\n",
    "    k_local=None,\n",
    "    key_local=\"X_pca\",\n",
    "    k_anchor=5,\n",
    "    key_anchor=\"X\",\n",
    "    dim_red=\"cca\",\n",
    "    max_cc_cells=100000,\n",
    "    k_score=30,\n",
    "    k_filter=min(200, min_sample),\n",
    "    scale1=False,\n",
    "    scale2=False,\n",
    "    n_components=n_cca_components,\n",
    "    n_features=100,\n",
    "    alignments=None,#[[[0], [1]]],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bb99f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "corrected = integrator.integrate(\n",
    "    key_correct=\"X_pca\",\n",
    "    row_normalize=True,\n",
    "    k_weight=50,\n",
    "    sd=1,\n",
    "    alignments=None, #[[[0], [1]]],\n",
    ")\n",
    "\n",
    "adata_merge.obsm[\"X_pca_integrate\"] = np.concatenate(corrected)\n",
    "# f=open(\"integrator.pkl\",'wb')\n",
    "# pickle.dump(integrator,f,True)\n",
    "# f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2587b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "if key_to_transfer is None:\n",
    "    key_to_transfer=[ref_cell_type]\n",
    "if isinstance(key_to_transfer, str): \n",
    "    key_to_transfer = [key_to_transfer]\n",
    "key_to_transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57b062bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "label_transfer=integrator.label_transfer(\n",
    "    ref=[0],qry=[1],categorical_key=key_to_transfer,continuous_key=None,\n",
    "    key_dist='X_pca', k_weight=k_weight_transfer\n",
    ")\n",
    "adata_merge.uns['label_transfer']=label_transfer\n",
    "for key in label_transfer:\n",
    "    labels = label_transfer[key].columns.tolist()\n",
    "    adata_merge.obs[f'infer_{key}'] = adata_merge.obs.index.to_series().map(\n",
    "        label_transfer[key].apply(lambda x: labels[np.argmax(x)], axis=1).to_dict())\n",
    "    adata_merge.obs[f'infer_{key}_prob'] = adata_merge.obs.index.to_series().map(\n",
    "        label_transfer[key].apply(lambda x: x[np.argmax(x)], axis=1).to_dict())\n",
    "    adata_merge.obs[key] = adata_merge.obs.apply(\n",
    "        lambda x: x[key] if not pd.isna(x[key]) else x[f'infer_{key}'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "785b60ed",
   "metadata": {},
   "source": [
    "### Harmony on embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03bad336",
   "metadata": {},
   "outputs": [],
   "source": [
    "sce.pp.harmony_integrate(adata_merge,\n",
    "                     key='Modality',basis='X_pca_integrate',\n",
    "                     nclust=50,max_iter_harmony=30) #X_pca_harmony"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbc48849",
   "metadata": {},
   "source": [
    "### tSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e579b14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 hour to run 4M + 3K cell\n",
    "tsne(adata_merge, obsm='X_pca_harmony',n_jobs=cpu)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f628c795",
   "metadata": {},
   "source": [
    "### UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b651a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 15 min to run 4M + 3K cell\n",
    "sc.pp.neighbors(adata_merge, use_rep='X_pca_harmony')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415a0771",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2 hours to run 4M + 3K cell\n",
    "print(resolution)\n",
    "sc.tl.leiden(adata_merge, resolution=resolution)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0642d637",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2 hours to run 4M + 3K cell\n",
    "# 4 hours to run 4M + 3K cell if using spectral init, the init step is very slow\n",
    "min_dist = max(0.1, 1 - adata_merge.shape[0] / 60000)\n",
    "try:\n",
    "    sc.tl.paga(adata_merge, groups=ref_cell_type) #leiden\n",
    "    sc.pl.paga(adata_merge, plot=False)\n",
    "    sc.tl.umap(adata_merge, min_dist=min_dist, init_pos='paga')\n",
    "except Exception:\n",
    "    print('Init with PAGA failed, use default spectral init')\n",
    "    sc.tl.umap(adata_merge, min_dist=min_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10cc1d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merge.write_h5ad(integrator_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "520287e3",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ee6852",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merge = ad.read_h5ad(integrator_path)\n",
    "adata_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f46be62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO: Change the modality to be passed in parameters\n",
    "ref_vc=adata_merge.obs.query(f\"Modality=='{mc_batch_key}'\").groupby('leiden')[ref_cell_type].value_counts(normalize=True).sort_values(ascending=False).reset_index()\n",
    "spatial_vc=adata_merge.obs.query(f\"Modality=='{spatial_batch_key}'\").groupby('leiden')[spatial_cell_type].value_counts(normalize=True).sort_values(ascending=False).reset_index()\n",
    "\n",
    "# mc_vc=mc_vc.loc[mc_vc.proportion > 0.5]\n",
    "# rna_vc=rna_vc.loc[rna_vc.proportion > 0.5]\n",
    "\n",
    "ref_vc.drop_duplicates('leiden',keep='first',inplace=True)\n",
    "spatial_vc.drop_duplicates('leiden',keep='first',inplace=True)\n",
    "ref_vc.rename(columns={'proportion':f'{ref_cell_type}_proportion'},inplace=True)\n",
    "spatial_vc.rename(columns={'proportion':f'{spatial_cell_type}_proportion'},inplace=True)\n",
    "df_map=pd.concat([ref_vc.set_index('leiden'),spatial_vc.set_index('leiden')],axis=1)\n",
    "df_map[f'{ref_cell_type}_cell_count']=df_map.index.to_series().map(adata_merge.obs.query(f\"Modality=='{mc_batch_key}'\").groupby('leiden')[ref_cell_type].count()).astype(int)\n",
    "df_map[f'{spatial_cell_type}_cell_count']=df_map.index.to_series().map(adata_merge.obs.query(f\"Modality=='{spatial_batch_key}'\").groupby('leiden')[spatial_cell_type].count()).astype(int)\n",
    "df_map.to_csv(cell_mapping_out_path,sep='\\t')\n",
    "df_map.query(f\"`{ref_cell_type}_cell_count` > 100 & `{spatial_cell_type}_cell_count` > 100\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5f37c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_merge.obs['Modality'].unique())\n",
    "hue_cols=[ref_cell_type,spatial_cell_type]\n",
    "modalities=[mc_batch_key,spatial_batch_key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c77fbb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if ref_palette is not None:\n",
    "    if os.path.isfile(os.path.expanduser(ref_palette)) and os.path.exists(os.path.expanduser(ref_palette)): #excel\n",
    "        D=pd.read_excel(os.path.expanduser(ref_palette),\n",
    "                    sheet_name=None, index_col=0)\n",
    "        ref_colors=D[ref_cell_type].Hex.to_dict()\n",
    "    else: #column from adata.obs\n",
    "        obs=ref_adata.obs.copy()\n",
    "        ref_colors=obs.reset_index().loc[:,[ref_cell_type,ref_palette]].drop_duplicates().dropna().set_index(ref_cell_type)[ref_palette].to_dict()\n",
    "adata_merge.uns[f'{ref_cell_type}_colors']=[ref_colors.get(k,'grey') for k in adata_merge.obs[ref_cell_type].cat.categories.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dda3bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not spatial_palette is None:\n",
    "    if os.path.isfile(os.path.expanduser(spatial_palette)) and os.path.exists(os.path.expanduser(spatial_palette)): #excel\n",
    "        D=pd.read_excel(os.path.expanduser(spatial_palette),\n",
    "                    sheet_name=None, index_col=0)\n",
    "        spatial_colors=D[spatial_cell_type].Hex.to_dict()\n",
    "    else: #column from adata.obs\n",
    "        obs=spatial_adata.obs.copy()\n",
    "        spatial_colors=obs.reset_index().loc[:,[spatial_cell_type,spatial_palette]].drop_duplicates().dropna().set_index(spatial_cell_type)[spatial_palette].to_dict()\n",
    "if spatial_colors is None:\n",
    "    sc.pl.umap(adata_merge,color=[spatial_cell_type],cmap='jet',\n",
    "               ncols=2,wspace=0.25,show=False,vmin='p5',vmax='p95')\n",
    "    atac_colors={cluster:color for cluster,color in zip(adata_merge.obs[spatial_cell_type].cat.categories.tolist(),adata_merge.uns[f'{spatial_cell_type}_colors'])}\n",
    "adata_merge.uns[f'{spatial_cell_type}_colors']=[spatial_colors.get(k,'grey') for k in adata_merge.obs[spatial_cell_type].cat.categories.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9872540c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dict={}\n",
    "for modality,hue_col in zip(modalities,hue_cols):\n",
    "    adata=adata_merge[adata_merge.obs['Modality'] == modality]\n",
    "    with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):    \n",
    "        sc.pl.tsne(adata, color=[hue_col], \n",
    "                   wspace=0.8)\n",
    "        sc.pl.umap(adata, color=[hue_col], \n",
    "                   wspace=0.8)\n",
    "    adata_dict[modality]=adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "787dbdd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for coord_base in ['tsne', 'umap']:\n",
    "    fig, axes = plt.subplots(nrows=1,ncols=2,figsize=(8, 3.5),dpi=300,constrained_layout=True)\n",
    "    for ax,col,modality in zip(axes,hue_cols,modalities):\n",
    "        adata=adata_dict[modality]\n",
    "        colors={cluster:color for cluster,color in zip(adata.obs[col].cat.categories.tolist(),adata.uns[f'{col}_colors'])}\n",
    "        ncol=2 if len(colors)>40 else 1\n",
    "        categorical_scatter(data=adata_merge[adata_merge.obs['Modality'] != modality],\n",
    "                            coord_base=coord_base,\n",
    "                            max_points=None,hue=None,\n",
    "                            scatter_kws=dict(color='lightgrey'),ax=ax)\n",
    "        categorical_scatter(data=adata,\n",
    "                            ax=ax,coord_base=coord_base,\n",
    "                            hue=col,\n",
    "                            palette=colors,\n",
    "                            max_points=None,\n",
    "                            # dodge_text=True,dodge_kws={\n",
    "                            #          \"force_points\": (15, 1),\n",
    "                            #           'autoalign':'y'},\n",
    "                            show_legend=True,\n",
    "                            legend_kws=dict(\n",
    "                                ncol=ncol,loc='upper left',bbox_to_anchor=(1,1),\n",
    "                                borderpad=0.4, # pad between marker (text) and border\n",
    "                                labelspacing=0.2, #The vertical space between the legend entries, in font-size units\n",
    "                                handleheight=0.5, #The height of the legend handles, in font-size units.\n",
    "                                handletextpad=0.2, # The pad between the legend handle (marker) and text, in font-size units.\n",
    "                                borderaxespad=0.2, # The pad between the Axes and legend border, in font-size units\n",
    "                                columnspacing=0.2, #The spacing between columns, in font-size units\n",
    "                                fontsize=5.5,title_fontsize=6\n",
    "                                )\n",
    "                           )\n",
    "        ax.set_title(modality,fontsize=14)\n",
    "    plt.subplots_adjust(hspace=0,wspace=0,left=0,right=0)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(f\"{image_path}/seurat.integrated.{coord_base}_with_legend.pdf\",dpi=300,bbox_inches='tight') #\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71faad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):    \n",
    "        sc.pl.tsne(adata_merge, color=['leiden'], \n",
    "                   wspace=0.8)\n",
    "        sc.pl.umap(adata_merge, color=['leiden'], \n",
    "                   wspace=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cd2b7bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "for coord_base in ['tsne', 'umap']:\n",
    "    #fig, axes = plt.subplots(nrows=1,ncols=3,figsize=(12, 4),dpi=300,constrained_layout=True)\n",
    "    col=\"leiden\"\n",
    "    fig, ax = plt.subplots(figsize=(5,4), dpi=300)\n",
    "    colors={cluster:color for cluster,color in zip(adata_merge.obs[col].cat.categories.tolist(),adata_merge.uns[f'{col}_colors'])}\n",
    "    categorical_scatter(data=adata_merge,\n",
    "                        ax=ax,coord_base=coord_base,\n",
    "                        hue=col,text_anno=adata.obs[col].cat.codes.map(str),#text_anno=col,\n",
    "                        palette=colors,\n",
    "                        max_points=None,\n",
    "                        text_kws=dict(color='white',fontweight=\"bold\",fontsize=4,\n",
    "                                     bbox=dict(facecolor=colors,boxstyle='circle', #ellipse, round\n",
    "                                               edgecolor='white',fill=True,linewidth=0.5,alpha=0.75)),\n",
    "                        luminance=0.48,\n",
    "                        # dodge_text=True,dodge_kws={\n",
    "                        #          \"force_points\": (15, 1),\n",
    "                        #           'autoalign':'y'},\n",
    "                        show_legend=True,#legend_kws=dict(ncol=1)\n",
    "                       )\n",
    "    ax.set_title(col,fontsize=14)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(f\"{image_path}/seurat.integrated.leiden.{coord_base}.pdf\",dpi=300,bbox_inches='tight') #\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2ed9b07",
   "metadata": {},
   "outputs": [],
   "source": [
    "for coord_base in ['tsne', 'umap']:\n",
    "    #fig, axes = plt.subplots(nrows=1,ncols=3,figsize=(12, 4),dpi=300,constrained_layout=True)\n",
    "    for col,modality in zip(hue_cols,modalities):\n",
    "        fig, ax = plt.subplots(figsize=(4.5, 4), dpi=300)\n",
    "        adata=adata_dict[modality]\n",
    "        colors={cluster:color for cluster,color in zip(adata.obs[col].cat.categories.tolist(),adata.uns[f'{col}_colors'])}\n",
    "        ncol=2 if len(colors)>20 else 1\n",
    "        categorical_scatter(data=adata_merge[adata_merge.obs['Modality'] != modality],\n",
    "                            coord_base=coord_base,\n",
    "                            max_points=None,hue=None,\n",
    "                            scatter_kws=dict(color='lightgrey'),ax=ax)\n",
    "        categorical_scatter(data=adata,\n",
    "                            ax=ax,coord_base=coord_base,\n",
    "                            hue=col,text_anno=col,\n",
    "                            palette=colors,\n",
    "                            max_points=None,\n",
    "                            text_kws=dict(color='black',fontweight=\"bold\",fontsize=4,#fontstretch='ultra-condensed',\n",
    "                                     bbox=dict(facecolor=colors,boxstyle='round', #ellipse, round\n",
    "                                               edgecolor='white',fill=True,linewidth=0.5,alpha=0.75)),\n",
    "                            luminance=0.48,\n",
    "                            # dodge_text=True,dodge_kws={\n",
    "                            #          \"force_points\": (15, 1),\n",
    "                            #           'autoalign':'y'},\n",
    "                            show_legend=False,legend_kws=dict(ncol=ncol,fontsize=5.5))\n",
    "        ax.set_title(modality,fontsize=14)\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(f\"{image_path}/seurat.integrated.{modality}.{coord_base}_without_legend.pdf\",dpi=300,bbox_inches='tight') #\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e4e646a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for coord_base in ['tsne', 'umap']:\n",
    "    #fig, axes = plt.subplots(nrows=1,ncols=3,figsize=(12, 4),dpi=300,constrained_layout=True)\n",
    "    for col,modality in zip(hue_cols,modalities):\n",
    "        fig, ax = plt.subplots(figsize=(5,4), dpi=300)\n",
    "        adata=adata_dict[modality]\n",
    "        ct2code=adata.obs.assign(code=adata.obs[col].cat.codes).loc[:,[col,'code']].drop_duplicates().set_index(col).code.to_dict()\n",
    "        id_colors={str(ct2code[cluster]):color for cluster,color in zip(adata.obs[col].cat.categories.tolist(),adata.uns[f'{col}_colors'])}\n",
    "        colors={f\"{ct2code[cluster]}: {cluster}\":color for cluster,color in zip(adata.obs[col].cat.categories.tolist(),adata.uns[f'{col}_colors'])}\n",
    "        ncol=2 if len(colors)>40 else 1\n",
    "        adata.obs[f'code_{col}']=adata.obs[col].apply(lambda x:f\"{ct2code[x]}: {x}\")\n",
    "        categorical_scatter(data=adata_merge[adata_merge.obs['Modality'] != modality],\n",
    "                            coord_base=coord_base,\n",
    "                            max_points=None,hue=None,\n",
    "                            scatter_kws=dict(color='lightgrey'),ax=ax)\n",
    "        categorical_scatter(data=adata,\n",
    "                            ax=ax,coord_base=coord_base,\n",
    "                            hue=f'code_{col}',text_anno=adata.obs[col].cat.codes.map(str),#text_anno=col,\n",
    "                            palette=colors,\n",
    "                            max_points=None,#text_anno_palette=id_colors,\n",
    "                            #text_anno=mc_data.obs[mc_col].replace(mask_dict_mc),\n",
    "                            text_kws=dict(color='white',fontweight=\"bold\",fontsize=4,\n",
    "                                     bbox=dict(facecolor=id_colors,boxstyle='round', #ellipse, round,circle\n",
    "                                               edgecolor='white',fill=True,linewidth=0.5,alpha=0.75)),\n",
    "                            luminance=0.48,\n",
    "                            # dodge_text=True,dodge_kws={\n",
    "                            #          \"force_points\": (15, 1),\n",
    "                            #           'autoalign':'y'},\n",
    "                            show_legend=True,\n",
    "                            legend_kws=dict(\n",
    "                                ncol=ncol,loc='upper left',bbox_to_anchor=(1,1),\n",
    "                                borderpad=0.4, # pad between marker (text) and border\n",
    "                                labelspacing=0.2, #The vertical space between the legend entries, in font-size units\n",
    "                                handleheight=0.5, #The height of the legend handles, in font-size units.\n",
    "                                handletextpad=0.2, # The pad between the legend handle (marker) and text, in font-size units.\n",
    "                                borderaxespad=0.2, # The pad between the Axes and legend border, in font-size units\n",
    "                                columnspacing=0.2, #The spacing between columns, in font-size units\n",
    "                                fontsize=5.5,title_fontsize=6\n",
    "                                )\n",
    "                           )\n",
    "        ax.set_title(modality,fontsize=14)\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(f\"{image_path}/seurat.integrated.{modality}.{coord_base}_with_code.pdf\",dpi=300,bbox_inches='tight') #\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32b6bfdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_adata=adata_merge[adata_merge.obs['Modality'] == spatial_batch_key].to_memory()\n",
    "if spatial_cell_type==ref_cell_type:\n",
    "    new_spatial_cell_type=f'SPATIAL.{spatial_cell_type}'\n",
    "    # use_adata.obs.rename(columns={rna_cell_type:new_rna_cell_type},inplace=True)\n",
    "    spatial_cell_type=new_spatial_cell_type\n",
    "key=ref_cell_type\n",
    "df=use_adata.obs.loc[:,[spatial_cell_type,f'infer_{key}',f'infer_{key}_prob']]\n",
    "df=df.loc[df[f'infer_{key}_prob']>=0.5]\n",
    "D=df[f'infer_{key}'].to_dict()\n",
    "use_adata.obs[key]=use_adata.obs_names.map(D)\n",
    "use_obs=df.index.tolist()\n",
    "with plt.rc_context({\"figure.figsize\": (4, 4), \"figure.dpi\": (100)}):\n",
    "    sc.pl.tsne(use_adata[use_obs,:], color=[f'infer_{key}',f'infer_{key}_prob'],\n",
    "               wspace=0.8)\n",
    "    sc.pl.umap(use_adata[use_obs,:], color=[f'infer_{key}',f'infer_{key}_prob'],\n",
    "               wspace=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19af206b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=use_adata.obs.loc[:,[spatial_cell_type,f'infer_{key}']].value_counts().reset_index()\n",
    "D=df.groupby(spatial_cell_type)['count'].sum()\n",
    "df['N']=df[spatial_cell_type].map(D).astype(int)\n",
    "df['fraction']=df['count'] / df.N\n",
    "data=df.pivot(index=spatial_cell_type,columns=f'infer_{key}',values='fraction')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28abeed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rows=data.index.to_series().to_frame()\n",
    "cols=data.columns.tolist()\n",
    "max_idx=np.argmax(data.fillna(0).values,axis=1)\n",
    "df_rows['GROUP']=[cols[i] for i in max_idx]\n",
    "use_rows=[]\n",
    "for col in data.columns.tolist():\n",
    "    df1=df_rows.loc[df_rows.GROUP==col]\n",
    "    if df1.shape[0]==0:\n",
    "        continue\n",
    "    use_rows.extend(df1[spatial_cell_type].unique().tolist())\n",
    "df_rows=df_rows.loc[use_rows]\n",
    "ct2code=use_adata.obs.assign(code=use_adata.obs[spatial_cell_type].cat.codes).loc[:,[spatial_cell_type,'code']].drop_duplicates().set_index(spatial_cell_type).code.to_dict()\n",
    "df_rows['Label']=df_rows[spatial_cell_type].apply(lambda x:f\"{ct2code[x]}: {x}\")\n",
    "df_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77787558",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_rows.groupby('GROUP').apply(lambda x:x.Label.apply(lambda x:int(x.split(':')[0])).tolist()).to_dict())\n",
    "print(df_rows.groupby('GROUP').apply(lambda x:x.Label.apply(lambda x:x.split(':')[1].strip()).tolist()).to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf1616e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PyComplexHeatmap import *\n",
    "row_ha=HeatmapAnnotation(\n",
    "    label=anno_label(df_rows.Label,colors='black',relpos=(0,0.5)),\n",
    "    axis=0,orientation='right',\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(24,12))\n",
    "ClusterMapPlotter(\n",
    "    data.loc[df_rows.index.tolist()],row_cluster=False,col_cluster=False,cmap='Reds',\n",
    "    right_annotation=row_ha,row_split=df_rows['GROUP'],row_split_gap=0.5,\n",
    "    row_split_order=df_rows['GROUP'].unique().tolist(),\n",
    "    show_rownames=False,show_colnames=True,yticklabels=True,xticklabels=True,\n",
    "    xticklabels_kws=dict(labelrotation=-60,labelcolor='blue',labelsize=10),\n",
    "    yticklabels_kws=dict(labelcolor='red',labelsize=10),\n",
    "    annot=True,fmt='.2g',linewidth=0.05,linecolor='gold',linestyle='-:',\n",
    "    label='fraction',legend_kws=dict(extend='both',extendfrac=0.1),\n",
    "    xlabel=ref_cell_type,ylabel=spatial_cell_type,\n",
    "    xlabel_kws=dict(color='blue',fontsize=14,labelpad=5),xlabel_side='top',\n",
    "    ylabel_kws=dict(color='red',fontsize=14,labelpad=5), #increace labelpad manually using labelpad (points)\n",
    "    # xlabel_bbox_kws=dict(facecolor='green'),\n",
    "    # ylabel_bbox_kws=dict(facecolor='chocolate',edgecolor='red'),|\n",
    "    # standard_scale=0,\n",
    ")\n",
    "plt.savefig(f\"{image_path}/confusion_matrix.pdf\",dpi=300,bbox_inches='tight') #\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f99a854",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_col = mc_cell_type\n",
    "cluster_col = \"leiden\"\n",
    "\n",
    "use_adata = adata_merge[adata_merge.obs['Modality'] == mc_batch_key].copy()\n",
    "vc = use_adata.obs.loc[:, [cluster_col, ref_col]].value_counts().reset_index()\n",
    "D = vc.groupby(cluster_col)['count'].sum()\n",
    "vc['N']=vc[cluster_col].map(D).astype(int)\n",
    "vc['fraction']=vc['count']/vc['N']\n",
    "data = vc.pivot(index=cluster_col, columns=ref_col, values='fraction')\n",
    "data.head()\n",
    "\n",
    "df_rows=data.index.to_series().to_frame()\n",
    "cols=data.columns.tolist()\n",
    "max_idx=np.argmax(data.fillna(0).values,axis=1)\n",
    "df_rows[\"GROUP\"]=[cols[i] for i in max_idx]\n",
    "use_rows=[]\n",
    "for col in data.columns.tolist(): \n",
    "    df1=df_rows.loc[df_rows['GROUP']==col]\n",
    "    if df1.shape[0]==0:\n",
    "        continue\n",
    "    use_rows.extend(df1[cluster_col].unique().tolist())\n",
    "df_rows=df_rows.loc[use_rows]\n",
    "ct2code=use_adata.obs.assign(code=use_adata.obs[cluster_col].cat.codes).loc[:,[cluster_col,'code']].drop_duplicates().set_index(cluster_col).code.to_dict()\n",
    "# df_rows['Label']=df_rows[cluster_col].apply(lambda x: f\"{ct2code[x]}: {x}\")\n",
    "ret = []\n",
    "for x in df_rows[cluster_col].tolist():\n",
    "    ret.extend([f\"{ct2code[x]}: {x}\"])\n",
    "df_rows['Label']=ret\n",
    "df_rows.head()\n",
    "\n",
    "# Plot\n",
    "row_ha=HeatmapAnnotation(\n",
    "    label=anno_label(df_rows.Label,colors='black',relpos=(0,0.5)),\n",
    "    axis=0,orientation='right',\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(24,12))\n",
    "ClusterMapPlotter(\n",
    "    data.loc[df_rows.index.tolist()],row_cluster=False,col_cluster=False,cmap='Reds',\n",
    "    right_annotation=row_ha,row_split=df_rows['GROUP'],row_split_gap=0.5,\n",
    "    row_split_order=df_rows['GROUP'].unique().tolist(),\n",
    "    show_rownames=False,show_colnames=True,yticklabels=True,xticklabels=True,\n",
    "    xticklabels_kws=dict(labelrotation=-60,labelcolor='blue',labelsize=10),\n",
    "    yticklabels_kws=dict(labelcolor='red',labelsize=10),\n",
    "    annot=True,fmt='.2g',linewidth=0.05,linecolor='gold',linestyle='-:',\n",
    "    label='fraction',legend_kws=dict(extend='both',extendfrac=0.1),\n",
    "    xlabel=ref_col,ylabel=cluster_col,\n",
    "    xlabel_kws=dict(color='blue',fontsize=14,labelpad=5),xlabel_side='top',\n",
    "    ylabel_kws=dict(color='red',fontsize=14,labelpad=5), #increace labelpad manually using labelpad (points)\n",
    "    # xlabel_bbox_kws=dict(facecolor='green'),\n",
    "    # ylabel_bbox_kws=dict(facecolor='chocolate',edgecolor='red'),|\n",
    "    # standard_scale=0,\n",
    ")\n",
    "plt.savefig(f\"{image_path}/integrated_confusion_matrix.pdf\",dpi=300,bbox_inches='tight')\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "preprocessing",
   "language": "python",
   "name": "preprocessing"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
