{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff58d988",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83cfc3ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "        \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", \"STH\", \"MGM\", \"GP\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06558d1a",
   "metadata": {},
   "source": [
    "### add_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c0c9c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "ret = subprocess.run(\"ll\", shell=True, capture_output=True)\n",
    "ret.stdout.decode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b65255de",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/add_metadata.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/00_add_metadata.ipynb\"\n",
    "SUFFIX = \"CPSfilt\" # \"cellpose_SAM_filt\" # \"proseg_fv38_filt\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "\n",
    "add_metadata_ret = {}\n",
    "for exp in experiments:\n",
    "    execute_out_dir = run_dir / exp / \"notebooks\"\n",
    "    script_out_dir = run_dir / exp / \"slurm_scripts\"\n",
    "    execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    script_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = script_out_dir / f\"add_metadata_cp.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f:\n",
    "        content = bash_content.format(EXP=exp, SUFFIX=SUFFIX)\n",
    "        f.write(content)\n",
    "    ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "    add_metadata_ret[exp] = ret.stdout.decode()\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43da34a5",
   "metadata": {},
   "source": [
    "### Subclass scVIVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7779866",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclass_scviva_ret['PU'].strip().split(\" \")[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91c3c42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/subclass_scviva.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/00_subclass_scviva.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "subclass_scviva_ret = {}\n",
    "for exp in experiments: \n",
    "    execute_out_dir = run_dir / exp / \"notebooks\"\n",
    "    script_out_dir = run_dir / exp / \"slurm_scripts\"\n",
    "    execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    script_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = script_out_dir / f\"subclass_scviva.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f: \n",
    "        content = bash_content.format(EXP=exp)\n",
    "        f.write(content)\n",
    "    ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "    subclass_scviva_ret[exp] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "    print(ret.stdout.decode().strip())\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5deb23",
   "metadata": {},
   "source": [
    "### Subclass Integration (wubin) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1120e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c364c847",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/subclass_integration.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "for exp in experiments: \n",
    "    ref_exp = EXP_TO_REF_PATH[exp]\n",
    "    for _execute_dir in [\"notebooks/base_r1_subclass\", \"notebooks/base_r2_subclass\", \"notebooks/scviva_r1_subclass\", \"notebooks/scviva_r2_subclass\"]:\n",
    "        execute_out_dir = run_dir / exp / _execute_dir\n",
    "        execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    script_out_dir = run_dir / exp / \"slurm_scripts\"\n",
    "    script_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = script_out_dir / f\"subclass_integration.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f: \n",
    "        content = bash_content.format(EXP=exp, REF_EXP=ref_exp)\n",
    "        f.write(content)\n",
    "    # os.system(f\"sbatch {out_script}\")\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4d24556",
   "metadata": {},
   "source": [
    "## Neighborhood Annot (SPIDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "362fb185",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f2e9ae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/neihborhood_annot.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_neighborhood_annot.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    execute_out_dir = run_dir / exp / \"notebooks\"\n",
    "    script_out_dir = run_dir / exp / \"slurm_scripts\"\n",
    "    execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    script_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = script_out_dir / f\"neighborhood_annot.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f: \n",
    "        content = bash_content.format(EXP=exp, REF_EXP=EXP_TO_REF_PATH[exp])\n",
    "        f.write(content)\n",
    "    ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "    nb_annot_ret[exp] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "    print(ret.stdout.decode().strip())\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c6f6721",
   "metadata": {},
   "source": [
    "## Subclass Integration (SPIDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa7b6c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# experiments = [\"PU\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c537198a",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f19558a",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/subclass_annot.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_subclass_annot.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "for exp in experiments: \n",
    "    execute_out_dir = run_dir / exp / \"notebooks\"\n",
    "    script_out_dir = run_dir / exp / \"slurm_scripts\"\n",
    "    execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    script_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = script_out_dir / f\"subclass_annot.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f: \n",
    "        content = bash_content.format(EXP=exp, REF_EXP=EXP_TO_REF_PATH[exp])\n",
    "        f.write(content)\n",
    "    os.system(f\"sbatch {out_script}\")\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de41812a",
   "metadata": {},
   "source": [
    "## Group Annotations (SPIDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4a65c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae4e177a",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "bash_template =\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/bash_template.sh\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4609ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(bash_template, \"r\") as f: \n",
    "    bash_content = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d53446",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(bash_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "106f4e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# unknown_template = \"\"\"\n",
    "# pixi run \\\\\n",
    "#     papermill \\\\\n",
    "#     /home/x-aklein2/projects/aklein/BICAN/spida_dev/analysis/annotations/template/cluster_unknown.ipynb \\\\\n",
    "#     {run_dir}/{EXP}/cluster_unknown.ipynb \\\\\n",
    "#     -p EXPERIMENT {EXP}\n",
    "# \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1589332",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_group_annot.ipynb \\\\\n",
    "    {run_dir}/{EXP}/notebooks/group/group_annot_{subclass_filename}.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\"\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef627085",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", \"STH\", \"MGM\", \"GP\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d164fa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\"ALL_RNA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec71522",
   "metadata": {},
   "outputs": [],
   "source": [
    "for exp in experiments:\n",
    "    ref_exp = EXP_TO_REF_PATH[exp]\n",
    "    read_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{exp}/{exp}.h5ad\"\n",
    "    adata = ad.read_h5ad(read_path)\n",
    "\n",
    "    run = bash_content.format(EXP=exp)\n",
    "    # run += unknown_template.format(EXP=exp, run_dir=run_dir)\n",
    "\n",
    "    subclasses = adata.obs['c2c_allcools_label_Subclass'].unique().tolist()\n",
    "    for subclass in subclasses:\n",
    "        if subclass == \"unknown\":\n",
    "            continue\n",
    "        subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "        subclass_formatted = subclass_template.format(\n",
    "            run_dir=run_dir,\n",
    "            EXP=exp,\n",
    "            REF_EXP=ref_exp,\n",
    "            SUBCLASS=subclass,\n",
    "            subclass_filename=subclass_filename\n",
    "        )\n",
    "        run += subclass_formatted\n",
    "\n",
    "    mk_path = run_dir / exp / \"notebooks/group\"\n",
    "    mk_path.mkdir(parents=True, exist_ok=True)\n",
    "    write_path = run_dir / exp / \"slurm_scripts/group_annot.sh\"\n",
    "    write_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "    with open(write_path, \"w\") as f:\n",
    "        f.write(run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d98be52",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(run)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1a324e",
   "metadata": {},
   "source": [
    "## Group Annotations SPIDA (ALL) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03fad60e",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "bash_template =\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/bash_template.sh\"\n",
    "with open(bash_template, \"r\") as f: \n",
    "    bash_content = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c66a44af",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_group_annot.ipynb \\\\\n",
    "    {run_dir}/{EXP}/notebooks/group/group_annot_{subclass_filename}.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Combined.Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path \"{REF_ADATA_PATH}\" \\\\\n",
    "    -p spatial_adata_path \"{SPATIAL_ADATA_PATH}\" \\\\\n",
    "    -p save_path \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_{SUBSET_LEVEL}_{subclass_filename}.h5ad\" \\\\\n",
    "    -p integrator_path \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_{SUBSET_LEVEL}_{subclass_filename}_joint.h5ad\" \\\\\n",
    "    -p top_deg_genes 50 \\\\\n",
    "    -k preprocessing\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a203b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ref_path = \"/home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/all_downsampled_50pct_nonneurons.h5ad\"\n",
    "exp = \"ALL_RNA\"\n",
    "ref = \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_ALL/BG_pfv8_nn.h5ad\"\n",
    "subset_level = \"Subclass\"\n",
    "spatial_subset_col = \"Combined.Subclass\"\n",
    "# subset_name = \"Neurons\"\n",
    "# read_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{exp}/{exp}.h5ad\"\n",
    "adata = ad.read_h5ad(ref)\n",
    "run = bash_content.format(EXP=exp)\n",
    "\n",
    "subclasses = adata.obs['Combined.Subclass'].unique().tolist()\n",
    "for subclass in subclasses:\n",
    "    if subclass == \"unknown\":\n",
    "        continue\n",
    "    subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "    subclass_formatted = subclass_template.format(\n",
    "        run_dir=run_dir,\n",
    "        EXP=exp,\n",
    "        SUBCLASS=subclass,\n",
    "        SUBSET_LEVEL=subset_level,\n",
    "        REF_ADATA_PATH=rna_ref_path,\n",
    "        SPATIAL_ADATA_PATH=ref,\n",
    "        subclass_filename=subclass_filename,\n",
    "        \n",
    "    )\n",
    "    run += subclass_formatted\n",
    "\n",
    "mk_path = run_dir / exp / \"notebooks/group\"\n",
    "mk_path.mkdir(parents=True, exist_ok=True)\n",
    "write_path = run_dir / exp / \"slurm_scripts/group_annot_nn.sh\"\n",
    "write_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "with open(write_path, \"w\") as f:\n",
    "    f.write(run)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2f1339d",
   "metadata": {},
   "source": [
    "## mC Group Annotations (ALL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f073025",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute\")\n",
    "bash_template =\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/bash_template.sh\"\n",
    "with open(bash_template, \"r\") as f: \n",
    "    bash_content = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25436909",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -r mc_query \" CellClass == 'Nonneuron' \" \\\\\n",
    "subclass_template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_group_methyl_annot.ipynb \\\\\n",
    "    {run_dir}/{EXP}/notebooks/group/group_annot_{subclass_filename}.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p mc_subset_col \"Subclass\" \\\\\n",
    "    -p mc_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p spatial_subset_col \"Combined.Subclass\" \\\\\n",
    "    -p spatial_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p mc_adata_path \"{REF_ADATA_PATH}\" \\\\\n",
    "    -p spatial_adata_path \"{SPATIAL_ADATA_PATH}\" \\\\\n",
    "    -p integrator_path \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_{SUBSET_LEVEL}_{subclass_filename}_joint.h5ad\" \\\\\n",
    "    -p cell_mapping_out_path /home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{EXP}/{EXP}_{SUBSET_LEVEL}_{subclass_filename}_cell_mapping.tsv \\\\\n",
    "    -p topn 50 \\\\\n",
    "    -k preprocessing\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce5cb21",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ref_path = \"/home/x-aklein2/projects/aklein/BICAN/data/reference/mC/BG_BICAN.gene-CHN.h5ad\"\n",
    "exp = \"ALL_MC\"\n",
    "ref = \"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_ALL/BG_pfv8_neu.h5ad\"\n",
    "subset_level = \"Subclass\"\n",
    "spatial_subset_col = \"Combined.Subclass\"\n",
    "# subset_name = \"Neurons\"\n",
    "# read_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/annotation/BICAN_BG_{exp}/{exp}.h5ad\"\n",
    "adata = ad.read_h5ad(ref)\n",
    "run = bash_content.format(EXP=exp)\n",
    "\n",
    "subclasses = adata.obs['Combined.Subclass'].unique().tolist()\n",
    "for subclass in subclasses:\n",
    "    if subclass == \"unknown\":\n",
    "        continue\n",
    "    subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "    subclass_formatted = subclass_template.format(\n",
    "        run_dir=run_dir,\n",
    "        EXP=exp,\n",
    "        SUBCLASS=subclass,\n",
    "        SUBSET_LEVEL=subset_level,\n",
    "        REF_ADATA_PATH=rna_ref_path,\n",
    "        SPATIAL_ADATA_PATH=ref,\n",
    "        SPATIAL_SUBCLASS=spatial_subset_col,\n",
    "        subclass_filename=subclass_filename,\n",
    "        \n",
    "    )\n",
    "    run += subclass_formatted\n",
    "\n",
    "mk_path = run_dir / exp / \"notebooks/group\"\n",
    "mk_path.mkdir(parents=True, exist_ok=True)\n",
    "write_path = run_dir / exp / \"slurm_scripts/group_annot_neu.sh\"\n",
    "write_path.parent.mkdir(parents=True, exist_ok=True)\n",
    "with open(write_path, \"w\") as f:\n",
    "    f.write(run)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef8f2bd5",
   "metadata": {},
   "source": [
    "## Alignment - Ren Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7489aad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", \"SUBTH\", \"MGM\", \"GP\"\n",
    "]\n",
    "\n",
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"SUBTH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM\",\n",
    "}\n",
    "\n",
    "EXP_TO_REN_NAME = {\n",
    "    \"PU\": \"Pu\",\n",
    "    \"CAH\": \"CaH\",\n",
    "    \"CAB\": \"CaB\",\n",
    "    \"CAT\": \"CaT\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"SUBTH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c4ccb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_content = \"\"\"#!/bin/bash\n",
    "# FILENAME = ren_{EXP}_subclass_integration.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J REN_{EXP}_subclass_integration\n",
    "#SBATCH -p wholenode\n",
    "#SBATCH --time=1:00:00\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=128gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/alignment/ren_{EXP}_subclass_integration.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/alignment/ren_{EXP}_subclass_integration.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/integration/seurat_integration_v2.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/alignment/Subclass_Ren_{REN_EXP}/seurat_integration.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/REN/BICAN_BGC_RNA_fliterV2_20250806.h5ad \\\\\n",
    "    -p ref_query ' region == \"{REN_EXP}\" ' \\\\\n",
    "    -p ref_cell_type Subclass_name \\\\\n",
    "    -p ref_downsample 3000 \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/data/BICAN_BG_PFV8_annotated_v4.h5ad \\\\\n",
    "    -p spatial_query ' brain_region == \"{EXP}\" ' \\\\\n",
    "    -p spatial_cell_type Subclass \\\\\n",
    "    -p spatial_downsample 3000 \\\\\n",
    "    -p topn 50 \\\\\n",
    "    -p outdir /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/alignment/Subclass_Ren_{REN_EXP}/ \\\\\n",
    "    \n",
    "\n",
    "echo \"Done\"\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e9e7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/alignment\")\n",
    "\n",
    "for exp in experiments: \n",
    "    ref_exp = EXP_TO_REF_PATH[exp]\n",
    "    ren_exp = EXP_TO_REN_NAME[exp]\n",
    "    execute_out_dir = run_dir / f\"Subclass_Ren_{ren_exp}\"\n",
    "    execute_out_dir.mkdir(exist_ok=True, parents=True)\n",
    "    out_script = execute_out_dir / \"subclass_integration.sh\"\n",
    "\n",
    "    with open(out_script, \"w\") as f: \n",
    "        content = bash_content.format(EXP=exp, REF_EXP=ref_exp, REN_EXP=ren_exp)\n",
    "        f.write(content)\n",
    "    os.system(f\"sbatch {out_script}\")\n",
    "    print(f\"Submitted job for {exp}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80bd6ba5",
   "metadata": {},
   "source": [
    "# Region Donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f8a1c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", \"STH\", \"MGM\", \"GP\"\n",
    "]\n",
    "donors = [\n",
    "    \"UCI2424\", \"UCI4723\", \"UCI5224\", \"UWA7648\"\n",
    "]\n",
    "exclude = [(\"CAT\", \"UWA7648\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f896ec35",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bffb69b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/region_donor_annot.sh\"\n",
    "# run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor\")\n",
    "# input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_region_donor_annot.ipynb\"\n",
    "\n",
    "# with open(template, \"r\") as f: \n",
    "#     bash_content = f.read()\n",
    "# print(bash_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df624582",
   "metadata": {},
   "outputs": [],
   "source": [
    "# experiments = ['GP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ff3cb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/region_donor_annot.sh\"\n",
    "# run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor\")\n",
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/region_donor_annot_cpsam.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_region_donor_annot.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        out_dir = run_dir / f\"{exp}_{donor}\"\n",
    "        out_dir.mkdir(exist_ok=True, parents=True)\n",
    "        out_script = out_dir / \"region_donor_annot.sh\"\n",
    "        out_nb = out_dir / \"region_donor_annot.ipynb\"\n",
    "\n",
    "        with open(out_script, \"w\") as f: \n",
    "            content = bash_content.format(EXP=exp, REF_EXP=EXP_TO_REF_PATH[exp], DONOR=donor)\n",
    "            f.write(content)\n",
    "        ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "        nb_annot_ret[f\"{exp}_{donor}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "        print(ret.stdout.decode().strip())\n",
    "        print(f\"Submitted job for {exp} {donor}\")\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa3f078",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in nb_annot_ret.items():\n",
    "    subprocess.run(f\"scancel {v}\", shell=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec1e1a3a",
   "metadata": {},
   "source": [
    "# Region Donor Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c55ce55",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", \"STH\", \"MGM\", \"GP\"\n",
    "]\n",
    "donors = [\n",
    "    \"UCI2424\", \"UCI4723\", \"UCI5224\", \"UWA7648\"\n",
    "]\n",
    "exclude = [(\"CAT\", \"UWA7648\")]\n",
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ccf27fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_header = \"\"\"#!/bin/bash\n",
    "# FILENAME = group_annot.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J {EXP}_{DONOR}_group_annot\n",
    "#SBATCH -p wholenode\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=64gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor/{EXP}_{DONOR}_group.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor/{EXP}_{DONOR}_group.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f75e2edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor/{EXP}_{DONOR}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor/{EXP}_{DONOR}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor/{EXP}_{DONOR}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor/{EXP}_{DONOR}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor/{EXP}_{DONOR}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 50 \\\\\n",
    "    -k preprocessing\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c4d3098",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        ad_file = run_dir / f\"{exp}_{donor}/final.h5ad\"\n",
    "        adata = ad.read_h5ad(ad_file)\n",
    "        subclasses = adata.obs['Subclass'].unique().tolist()\n",
    "        run = bash_header.format(EXP=exp, DONOR=donor)\n",
    "        for subclass in subclasses:\n",
    "            if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                continue\n",
    "            subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "            subclass_formatted = template.format(\n",
    "                EXP=exp,\n",
    "                DONOR=donor,\n",
    "                REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                SUBCLASS=subclass,\n",
    "                subclass_filename=subclass_filename\n",
    "            )\n",
    "            run += subclass_formatted\n",
    "            out_dir = run_dir / f\"{exp}_{donor}/groups\" / subclass_filename\n",
    "            out_dir.mkdir(parents=True, exist_ok=True)\n",
    "        out_script = run_dir / f\"{exp}_{donor}/groups/group_annot.sh\"\n",
    "        with open(out_script, \"w\") as f:\n",
    "            f.write(run)\n",
    "        ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "        nb_annot_ret[f\"{exp}_{donor}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "        print(ret.stdout.decode().strip())\n",
    "        print(f\"Submitted job for {exp} {donor}\")\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8144f300",
   "metadata": {},
   "source": [
    "### CPSAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a73ccc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_header = \"\"\"#!/bin/bash\n",
    "# FILENAME = group_annot.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J {EXP}_{DONOR}_cp_group_annot\n",
    "#SBATCH -p wholenode\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=64gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_cpsam/{EXP}_{DONOR}_group.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_cpsam/{EXP}_{DONOR}_group.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c629e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam/{EXP}_{DONOR}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam/{EXP}_{DONOR}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam/{EXP}_{DONOR}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam/{EXP}_{DONOR}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam/{EXP}_{DONOR}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 50 \\\\\n",
    "    -k preprocessing_bp\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32627f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        ad_file = run_dir / f\"{exp}_{donor}/final.h5ad\"\n",
    "        adata = ad.read_h5ad(ad_file)\n",
    "        subclasses = adata.obs['Subclass'].unique().tolist()\n",
    "        run = bash_header.format(EXP=exp, DONOR=donor)\n",
    "        for subclass in subclasses:\n",
    "            if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                continue\n",
    "            subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "            subclass_formatted = template.format(\n",
    "                EXP=exp,\n",
    "                DONOR=donor,\n",
    "                REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                SUBCLASS=subclass,\n",
    "                subclass_filename=subclass_filename\n",
    "            )\n",
    "            run += subclass_formatted\n",
    "            out_dir = run_dir / f\"{exp}_{donor}/groups\" / subclass_filename\n",
    "            out_dir.mkdir(parents=True, exist_ok=True)\n",
    "        out_script = run_dir / f\"{exp}_{donor}/groups/group_annot.sh\"\n",
    "        with open(out_script, \"w\") as f:\n",
    "            f.write(run)\n",
    "        ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "        nb_annot_ret[f\"{exp}_{donor}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "        print(ret.stdout.decode().strip())\n",
    "        print(f\"Submitted job for {exp} {donor}\")\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0220e359",
   "metadata": {},
   "source": [
    "### scancel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7620d2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in nb_annot_ret.items():\n",
    "    subprocess.run(f\"scancel {v}\", shell=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3b2086",
   "metadata": {},
   "source": [
    "# Region Donor Lab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd204fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", # \"MGM\",# \"GP\", # \"STH\"\n",
    "]\n",
    "# experiments = [\n",
    "#     \"STH\"\n",
    "# ]\n",
    "experiments = [\n",
    "    \"GP\"\n",
    "]\n",
    "donors = [\n",
    "    \"UCI2424\", \"UCI4723\", \"UCI5224\", \"UWA7648\"\n",
    "]\n",
    "labs = [\"salk\", \"ucsd\"]\n",
    "exclude = [(\"CAT\", \"UWA7648\")]\n",
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b2f27c",
   "metadata": {},
   "source": [
    "## Proseg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b30c947",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/region_donor_lab_annot_gp.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_region_donor_lab_annot.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            out_dir = run_dir / f\"{exp}_{donor}_{lab}\"\n",
    "            out_dir.mkdir(exist_ok=True, parents=True)\n",
    "            out_script = out_dir / \"region_donor_lab_annot.sh\"\n",
    "            out_nb = out_dir / \"region_donor_lab_annot.ipynb\"\n",
    "\n",
    "        with open(out_script, \"w\") as f:\n",
    "            content = bash_content.format(EXP=exp, REF_EXP=EXP_TO_REF_PATH[exp], DONOR=donor, LAB1=labs[0], LAB2=labs[1])\n",
    "            f.write(content)\n",
    "        ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "        nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "        print(ret.stdout.decode().strip())\n",
    "        print(f\"Submitted job for {exp} {donor}\")\n",
    "        \n",
    "    #     break\n",
    "    # break\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6eef084",
   "metadata": {},
   "source": [
    "## CPSAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0512e4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/slurm_scripts/region_donor_lab_cpsam_annot_gp.sh\"\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2\")\n",
    "input_notebook = \"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/01_region_donor_lab_cpsam_annot.ipynb\"\n",
    "\n",
    "with open(template, \"r\") as f: \n",
    "    bash_content = f.read()\n",
    "# print(bash_content)\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            out_dir = run_dir / f\"{exp}_{donor}_{lab}\"\n",
    "            out_dir.mkdir(exist_ok=True, parents=True)\n",
    "            out_script = out_dir / \"region_donor_lab_annot.sh\"\n",
    "            out_nb = out_dir / \"region_donor_lab_annot.ipynb\"\n",
    "\n",
    "        with open(out_script, \"w\") as f:\n",
    "            content = bash_content.format(EXP=exp, REF_EXP=EXP_TO_REF_PATH[exp], DONOR=donor, LAB1=labs[0], LAB2=labs[1])\n",
    "            f.write(content)\n",
    "        ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "        nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "        print(ret.stdout.decode().strip())\n",
    "        print(f\"Submitted job for {exp} {donor}\")\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "490f6bd6",
   "metadata": {},
   "source": [
    "## scancel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c8b1f2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in nb_annot_ret.items():\n",
    "    subprocess.run(f\"scancel {v}\", shell=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60612cc9",
   "metadata": {},
   "source": [
    "# Region Donor Lab Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b11d427",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments = [\n",
    "    \"PU\", \"CAH\", \"CAB\", \"CAT\", \"NAC\", # \"MGM\", # \"GP\", \"STH\"\n",
    "]\n",
    "# experiments = [\n",
    "#     \"PU\", \"NAC\"\n",
    "# ]\n",
    "# experiments = [\n",
    "#     \"STH\", \"GP\"\n",
    "# ]\n",
    "donors = [\n",
    "    \"UCI2424\", \"UCI4723\", \"UCI5224\", \"UWA7648\"\n",
    "]\n",
    "labs = [\"salk\", \"ucsd\"]\n",
    "exclude = [(\"CAT\", \"UWA7648\")]\n",
    "EXP_TO_REF_PATH = {\n",
    "    \"PU\": \"PU\",\n",
    "    \"CAH\": \"CA\",\n",
    "    \"CAB\": \"CA\",\n",
    "    \"CAT\": \"CA\",\n",
    "    \"GP\": \"GP\",\n",
    "    \"STH\": \"STH\",\n",
    "    \"NAC\": \"NAC\",\n",
    "    \"MGM\": \"MGM1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e394f5fe",
   "metadata": {},
   "source": [
    "## Proseg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdd47c88",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_header = \"\"\"#!/bin/bash\n",
    "# FILENAME = rdl_group_annot.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J {EXP}_{DONOR}_{LAB}_group_annot\n",
    "#SBATCH -p shared\n",
    "#SBATCH --time=2:30:00\n",
    "#SBATCH --nodes=4\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=46gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26c03334",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp/{EXP}_{DONOR}_{LAB}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 150 \\\\\n",
    "    -p deg_type cef \\\\\n",
    "    -p joint_embedding_leiden_res 0.6 \\\\\n",
    "    -k preprocessing_bp\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "598d8f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_gp\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            ad_file = run_dir / f\"{exp}_{donor}_{lab}/final.h5ad\"\n",
    "            adata = ad.read_h5ad(ad_file)\n",
    "            subclasses = adata.obs['Subclass'].unique().tolist()\n",
    "            run = bash_header.format(EXP=exp, DONOR=donor, LAB=lab)\n",
    "            for subclass in subclasses:\n",
    "                if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                    continue\n",
    "                subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "                subclass_formatted = template.format(\n",
    "                    EXP=exp,\n",
    "                    DONOR=donor,\n",
    "                    LAB=lab,\n",
    "                    REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                    SUBCLASS=subclass,\n",
    "                    subclass_filename=subclass_filename\n",
    "                )\n",
    "                run += subclass_formatted\n",
    "                out_dir = run_dir / f\"{exp}_{donor}_{lab}/groups\" / subclass_filename\n",
    "                out_dir.mkdir(parents=True, exist_ok=True)\n",
    "            out_script = run_dir / f\"{exp}_{donor}_{lab}/groups/group_annot.sh\"\n",
    "            with open(out_script, \"w\") as f:\n",
    "                f.write(run)\n",
    "            ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "            nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "            print(ret.stdout.decode().strip())\n",
    "            print(f\"Submitted job for {exp} {donor} {lab}\")\n",
    "    #         break\n",
    "    #     break\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b134d8a",
   "metadata": {},
   "source": [
    "## CPSAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a523e7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_header = \"\"\"#!/bin/bash\n",
    "# FILENAME = rdl_cp_group_annot.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J cp_{EXP}_{DONOR}_{LAB}_group_annot\n",
    "#SBATCH -p shared\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH --nodes=4\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=46gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab_cpsam/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab_cpsam/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2294e9c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp/{EXP}_{DONOR}_{LAB}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 200 \\\\\n",
    "    -p deg_type cef \\\\\n",
    "    -p joint_embedding_leiden_res 0.65 \\\\\n",
    "    -k preprocessing_bp\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1123d750",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from_path = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_cpsam_lab\")\n",
    "# to_path = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam\")\n",
    "# for _exp_donor_lab_dir in from_path.glob(\"*\"):\n",
    "#     print(_exp_donor_lab_dir)\n",
    "#     for file in _exp_donor_lab_dir.glob(\"*\"):\n",
    "#         if file.is_dir():\n",
    "#             subprocess.run(f\"mv {file} {to_path}/{_exp_donor_lab_dir.name}/{file.name}\", shell=True)\n",
    "#         else:\n",
    "#             subprocess.run(f\"mv {file} {to_path}/{_exp_donor_lab_dir.name}/{file.name}\", shell=True)\n",
    "#     # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e97fbf4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cpsam_gp\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            ad_file = run_dir / f\"{exp}_{donor}_{lab}/final.h5ad\"\n",
    "            adata = ad.read_h5ad(ad_file)\n",
    "            subclasses = adata.obs['Subclass'].unique().tolist()\n",
    "            run = bash_header.format(EXP=exp, DONOR=donor, LAB=lab)\n",
    "            for subclass in subclasses:\n",
    "                if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                    continue\n",
    "                subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "                subclass_formatted = template.format(\n",
    "                    EXP=exp,\n",
    "                    DONOR=donor,\n",
    "                    LAB=lab,\n",
    "                    REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                    SUBCLASS=subclass,\n",
    "                    subclass_filename=subclass_filename\n",
    "                )\n",
    "                run += subclass_formatted\n",
    "                out_dir = run_dir / f\"{exp}_{donor}_{lab}/groups\" / subclass_filename\n",
    "                out_dir.mkdir(parents=True, exist_ok=True)\n",
    "            out_script = run_dir / f\"{exp}_{donor}_{lab}/groups/group_annot.sh\"\n",
    "            with open(out_script, \"w\") as f:\n",
    "                f.write(run)\n",
    "            ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "            nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "            print(ret.stdout.decode().strip())\n",
    "            print(f\"Submitted job for {exp} {donor} {lab}\")\n",
    "    #         break\n",
    "    #     break\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a8299a",
   "metadata": {},
   "source": [
    "## CPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf223fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_header = \"\"\"#!/bin/bash\n",
    "# FILENAME = rdl_cp_group_annot.sh \n",
    "\n",
    "#SBATCH -A mcb130189\n",
    "#SBATCH -J cps_{EXP}_{DONOR}_{LAB}_group_annot\n",
    "#SBATCH -p shared\n",
    "#SBATCH --time=3:00:00\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --mem=46gb\n",
    "#SBATCH -o /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH -e /home/x-aklein2/projects/aklein/BICAN/BG/logs/annot/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}_group.out\n",
    "#SBATCH --export=ALL\n",
    "\n",
    "export PATH=\"/home/x-aklein2/.pixi/bin:$PATH\"\n",
    "cd /anvil/projects/x-mcb130189/aklein/SPIDA\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e453f1a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2/{EXP}_{DONOR}_{LAB}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 200 \\\\\n",
    "    -p deg_type cef \\\\\n",
    "    -p joint_embedding_leiden_res 0.65 \\\\\n",
    "    -p run_harmony_joint_embeddings True \\\\\n",
    "    -k preprocessing\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f0b02b",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps2\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            ad_file = run_dir / f\"{exp}_{donor}_{lab}/final.h5ad\"\n",
    "            adata = ad.read_h5ad(ad_file)\n",
    "            subclasses = adata.obs['Subclass'].unique().tolist()\n",
    "            run = bash_header.format(EXP=exp, DONOR=donor, LAB=lab)\n",
    "            for subclass in subclasses:\n",
    "                if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                    continue\n",
    "                subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "                subclass_formatted = template.format(\n",
    "                    EXP=exp,\n",
    "                    DONOR=donor,\n",
    "                    LAB=lab,\n",
    "                    REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                    SUBCLASS=subclass,\n",
    "                    subclass_filename=subclass_filename\n",
    "                )\n",
    "                run += subclass_formatted\n",
    "                out_dir = run_dir / f\"{exp}_{donor}_{lab}/groups\" / subclass_filename\n",
    "                out_dir.mkdir(parents=True, exist_ok=True)\n",
    "            out_script = run_dir / f\"{exp}_{donor}_{lab}/groups/group_annot.sh\"\n",
    "            with open(out_script, \"w\") as f:\n",
    "                f.write(run)\n",
    "            ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "            nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "            print(ret.stdout.decode().strip())\n",
    "            print(f\"Submitted job for {exp} {donor} {lab}\")\n",
    "    #         break\n",
    "    #     break\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13926254",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Doing some V2 stuff for NAC and PU\n",
    "template = \"\"\"\n",
    "pixi run \\\\\n",
    "    papermill \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/template/02_region_donor_group_annot.ipynb \\\\\n",
    "    /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/integrate.ipynb \\\\\n",
    "    -p EXP {EXP} \\\\\n",
    "    -p REF_EXP {REF_EXP} \\\\\n",
    "    -p rna_subset_col \"Subclass\" \\\\\n",
    "    -p rna_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p qry_subset_col \"Subclass\" \\\\\n",
    "    -p qry_subset_value \"{SUBCLASS}\" \\\\\n",
    "    -p ref_adata_path /home/x-aklein2/projects/aklein/BICAN/data/reference/AIT/AIT_{REF_EXP}_filtered_2.h5ad \\\\\n",
    "    -p spatial_adata_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}/final.h5ad \\\\\n",
    "    -p save_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/final.h5ad \\\\\n",
    "    -p integrator_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/joint.h5ad \\\\\n",
    "    -p image_path /home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps/{EXP}_{DONOR}_{LAB}/groups/{subclass_filename}/figures \\\\\n",
    "    -p label_transfer_k 30 \\\\\n",
    "    -p top_deg_genes 200 \\\\\n",
    "    -p deg_type cef \\\\\n",
    "    -p joint_embedding_leiden_res 0.65 \\\\\n",
    "    -k preprocessing_bp\n",
    "\"\"\"\n",
    "\n",
    "run_dir = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/annotation/execute/region_donor_lab_cps\")\n",
    "\n",
    "nb_annot_ret = {}\n",
    "for exp in experiments: \n",
    "    for donor in donors:\n",
    "        if (exp, donor) in exclude:\n",
    "            print(f\"Skipping {exp} {donor}\")\n",
    "            continue\n",
    "        for lab in labs:\n",
    "            # ad_file = run_dir / f\"{exp}_{donor}_{lab}/final.h5ad\"\n",
    "            # adata = ad.read_h5ad(ad_file)\n",
    "            subclasses = [\"STR D1 MSN\", \"STR D2 MSN\"] # adata.obs['Subclass'].unique().tolist()\n",
    "            run = bash_header.format(EXP=exp, DONOR=donor, LAB=lab)\n",
    "            for subclass in subclasses:\n",
    "                if subclass == \"unknown\" or pd.isna(subclass):\n",
    "                    continue\n",
    "                subclass_filename = subclass.replace(\" \", \"_\").replace(\"/\", \"-\")\n",
    "                subclass_formatted = template.format(\n",
    "                    EXP=exp,\n",
    "                    DONOR=donor,\n",
    "                    LAB=lab,\n",
    "                    REF_EXP=EXP_TO_REF_PATH[exp],\n",
    "                    SUBCLASS=subclass,\n",
    "                    subclass_filename=subclass_filename\n",
    "                )\n",
    "                run += subclass_formatted\n",
    "                out_dir = run_dir / f\"{exp}_{donor}_{lab}/groups\" / subclass_filename\n",
    "                out_dir.mkdir(parents=True, exist_ok=True)\n",
    "            out_script = run_dir / f\"{exp}_{donor}_{lab}/groups/group_annot_v2.sh\"\n",
    "            with open(out_script, \"w\") as f:\n",
    "                f.write(run)\n",
    "            ret = subprocess.run(f\"sbatch {out_script}\", shell=True, capture_output=True)\n",
    "            nb_annot_ret[f\"{exp}_{donor}_{lab}\"] = ret.stdout.decode().strip().split(\" \")[-1]\n",
    "            print(ret.stdout.decode().strip())\n",
    "            print(f\"Submitted job for {exp} {donor} {lab}\")\n",
    "    #         break\n",
    "    #     break\n",
    "    # break\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8467c889",
   "metadata": {},
   "source": [
    "## scancel "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09abdd27",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in nb_annot_ret.items():\n",
    "    subprocess.run(f\"scancel {v}\", shell=True, capture_output=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7859be1c",
   "metadata": {},
   "source": [
    "# Notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b419949e",
   "metadata": {},
   "source": [
    "Jiang Notes: \n",
    "\n",
    "mitochondrial genes / ribosomal genes --> not well annotated genes that behave differently.\n",
    "Isoform specific genes. They look for the highest expressing isoforms of genes.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "preprocessing",
   "language": "python",
   "name": "preprocessing"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
