{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b4855e9b",
   "metadata": {},
   "source": [
    "The point of this notebook is to plot all the heatmaps + spatial plots of the SVGs associated with the Matrix Striosome Compartment switching. The figures from this notebook will go into spatial supp. figure 5 as well as the spatial main figure (Fig. 6). \n",
    "\n",
    "Author: Amit Klein \n",
    "Email: a3klein@ucsd.edu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bfe49c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "from scipy.stats import norm, spearmanr\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import geopandas as gpd\n",
    "from spida.utilities._ad_utils import normalize_adata\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "import seaborn as sns\n",
    "\n",
    "from spida.pl._utils import add_color_scheme\n",
    "from spida.pl import plot_categorical, plot_continuous, categorical_scatter, continuous_scatter\n",
    "from spida.utilities.sd_utils import _get_obs_or_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f24f07d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print versions of important packages\n",
    "print(f\"Python: {os.sys.version}\")\n",
    "print(f\"NumPy: {np.__version__}\")\n",
    "print(f\"Pandas: {pd.__version__}\")\n",
    "print(f\"Anndata: {ad.__version__}\")\n",
    "print(f\"Scanpy: {sc.__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050f51a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['savefig.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = (4, 4)\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = 'Arial'\n",
    "plt.rcParams['font.size'] = 8\n",
    "plt.rcParams['axes.facecolor'] = 'white'\n",
    "    \n",
    "plt.rcParams['image.interpolation'] = 'nearest'\n",
    "plt.rcParams['savefig.dpi'] = 300\n",
    "plt.rcParams['savefig.transparent'] = True\n",
    "plt.rcParams['savefig.bbox'] = 'tight'\n",
    "plt.rcParams['savefig.pad_inches'] = 0.01\n",
    "\n",
    "RASTERIZED = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e18c553",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_path = \"/home/x-aklein2/projects/aklein/BICAN/BG/data/BICAN_BG_CPS.h5ad\"\n",
    "adata = ad.read_h5ad(ad_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70acc54f",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_mch = ad.read_h5ad(\"/home/x-aklein2/projects/aklein/BICAN/BG/data/methylation_2/BG_mCH_Imp_SubR.h5ad\")\n",
    "for k, v in adata.uns.items(): \n",
    "    if \"colors\" in k or \"palette\" in k: \n",
    "        spatial_mch.uns[k] = v\n",
    "spatial_mch.obsm['spatial'] = spatial_mch.obs[['CENTER_X', 'CENTER_Y']].values\n",
    "spatial_mch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c01e9078",
   "metadata": {},
   "outputs": [],
   "source": [
    "image_path = Path(\"/home/x-aklein2/projects/aklein/BICAN/BG/images/figures/ms-svg3\")\n",
    "image_path.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8784e88c",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca00f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downsample Function\n",
    "def _downsample_reference(\n",
    "    ref_adata : ad.AnnData,\n",
    "    cluster_col : str,\n",
    "    max_cluster_size: int = 3000,\n",
    "    min_cluster_size: int = 0,\n",
    "):\n",
    "    \"\"\"\n",
    "    Remove clusters from the reference that have less than min_cluster_size cells.\n",
    "    Downsample larger clusters that have more than max_cluster_size cells.\n",
    "    \"\"\"\n",
    "    from spida.utilities._ad_utils import _downsample_ref_clusters, _remove_small_clusters\n",
    "    if min_cluster_size > 0: \n",
    "        ref_adata = _remove_small_clusters(ref_adata, cluster_col, min_cells=min_cluster_size)\n",
    "    if max_cluster_size > 0:\n",
    "        ref_adata = _downsample_ref_clusters(ref_adata, cluster_col, max_cells=max_cluster_size)\n",
    "    return ref_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa565e04",
   "metadata": {},
   "outputs": [],
   "source": [
    "def expr_heatmap(\n",
    "    adata,\n",
    "    df_genes,\n",
    "    n_genes=40,\n",
    "    gene_names=10,\n",
    "    heatmap_order_col='rho_axis',\n",
    "    groupby_replicate=True,\n",
    "    replicate_col=\"brain_region\",\n",
    "    replicate_palette=None,\n",
    "    replicate_order=None,\n",
    "    title=\"\",\n",
    "    filename = None,\n",
    "    out_path=None,\n",
    "    ylabel=\"Genes\",\n",
    "    xlabel=\"Cells\",\n",
    "    cmap=\"plasma\",\n",
    "    rasterized=True,\n",
    "    show=True, \n",
    "    save=False,\n",
    "    image_path=None,\n",
    "    min_max_quantiles=(0.02, 0.98),\n",
    "    vmin=0, vmax=1,\n",
    "    color_splits=None,\n",
    "):\n",
    "    \"\"\"Plot expression heatmap for given genes and cell type\"\"\"\n",
    "    import PyComplexHeatmap as pch\n",
    "    from matplotlib.colors import TwoSlopeNorm\n",
    "\n",
    "    if filename is None: \n",
    "        filename = f\"expr_heatmap_{title.replace(' ', '_')}\"\n",
    "\n",
    "    if replicate_palette is None:\n",
    "        replicate_palette = adata.uns.get(f\"{replicate_col}_palette\", None)\n",
    "\n",
    "    df_col = adata.obs[['Group', 'MS_NORM', 'MS_compartment']].copy()\n",
    "    if df_col.shape[0] > 50000:\n",
    "        df_col = df_col.sample(50000, random_state=42)\n",
    "    if groupby_replicate:\n",
    "        df_col[replicate_col] = adata.obs[replicate_col]\n",
    "        df_col = df_col.sort_values([replicate_col, 'MS_NORM'])\n",
    "    else:\n",
    "        df_col = df_col.sort_values('MS_NORM')\n",
    "\n",
    "    df_row = df_genes.iloc[:n_genes].copy()\n",
    "    df_row.drop_duplicates(subset='gene', keep='first', inplace=True)\n",
    "    if \"gene\" in df_row.columns:\n",
    "        df_row = df_row.set_index('gene')\n",
    "    toplot = df_row.index[:gene_names]\n",
    "    df_row['annot'] = [c if c in toplot else np.nan for c in df_row.index]\n",
    "    df_row = df_row.sort_values(heatmap_order_col, ascending=True)\n",
    "    # if \"gene\" in df_row.columns:\n",
    "    #     df_row = df_row.set_index('gene')\n",
    "\n",
    "    if hasattr(adata.X, \"toarray\"):\n",
    "        df_expr = adata.X.toarray()\n",
    "    else: \n",
    "        df_expr = adata.X\n",
    "    df_expr = pd.DataFrame(df_expr, index=adata.obs_names, columns=adata.var_names).T\n",
    "    df_expr = df_expr.loc[df_row.index, df_col.index]\n",
    "    df_expr_norm = (\n",
    "        df_expr\n",
    "        .subtract(df_expr.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "        .div(df_expr.quantile(min_max_quantiles[1], axis=1) - df_expr.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    )\n",
    "\n",
    "    ms_score_norm = TwoSlopeNorm(vmin=-1, vcenter=0, vmax=1)\n",
    "    col_ha = pch.HeatmapAnnotation(\n",
    "        # label=pch.anno_label(\n",
    "        #     df_col['MS_compartment'], merge=True, rotation=0, extend=True,\n",
    "        #     colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, \n",
    "        # ),\n",
    "        Brain_Region=pch.anno_simple(df_col[replicate_col], colors=replicate_palette, add_text=True, legend=False, height=3, text_kws={\"fontsize\":8, \"fontweight\":\"bold\"}),\n",
    "        Compartment=pch.anno_simple(df_col['MS_compartment'], colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, height=3),\n",
    "        MatStr_Score=pch.anno_simple(df_col['MS_NORM'], cmap=\"coolwarm_r\", norm=ms_score_norm, height=3),\n",
    "        verbose=1, axis=1, plot_legend=True, legend_gap=5,\n",
    "    )\n",
    "\n",
    "    left_ha = pch.HeatmapAnnotation(\n",
    "        label=pch.anno_label(\n",
    "            df_row['annot'], merge=True, rotation=0, extend=True,\n",
    "            colors=\"black\", relpos=(1, 0.5), \n",
    "            arrowprops = dict(visible=True,)\n",
    "        ),\n",
    "        # Genes=pch.anno_simple(df_row[0]),\n",
    "        verbose=1, axis=0\n",
    "    )\n",
    "\n",
    "    plt.figure(figsize=(8,4))\n",
    "    cm = pch.ClusterMapPlotter(\n",
    "        data=df_expr_norm,\n",
    "        top_annotation=col_ha,\n",
    "        left_annotation=left_ha,\n",
    "        row_cluster=False,\n",
    "        col_cluster=False,\n",
    "        col_split=df_col[replicate_col] if groupby_replicate else None,\n",
    "        row_dendrogram=False,\n",
    "        label=\"Expression\",\n",
    "        cmap=cmap,\n",
    "        rasterized=rasterized,\n",
    "        ylabel=ylabel,\n",
    "        xlabel=title,\n",
    "        xlabel_side=\"top\",\n",
    "        vmin=vmin, vmax=vmax,\n",
    "        col_split_order=replicate_order\n",
    "    )\n",
    "\n",
    "    # plot custom spines # For coloring the region in between the col splits as something other than white! \n",
    "    if color_splits is not None: \n",
    "        for i in range(cm.heatmap_axes.shape[0]-1):\n",
    "            for j in range(cm.heatmap_axes.shape[1]):\n",
    "                # if i != j:\n",
    "                #     continue\n",
    "                ax = cm.heatmap_axes[i][j]\n",
    "                for side in [\"right\"]:\n",
    "                    ax.spines[side].set_visible(True)\n",
    "                    ax.spines[side].set_color(color_splits)\n",
    "                    ax.spines[side].set_linewidth(2)\n",
    "\n",
    "    # plt.suptitle(title)\n",
    "    if save and image_path is not None:\n",
    "        print(\"saving\")\n",
    "        plt.savefig(image_path / f\"{filename}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "        plt.savefig(image_path / f\"{filename}.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "        # plt.savefig(image_path / f\"{filename}.svg\", dpi=300, bbox_inches=\"tight\")\n",
    "        # pass\n",
    "    if show: \n",
    "        plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa68599f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# WUBIN THIS IS WHERE THE FUNCTION THAT DOES THE ACTUAL PLOTTING LIVES. \n",
    "def joint_expr_heatmap(\n",
    "    adata_rna,\n",
    "    adata_mch,\n",
    "    df_genes,\n",
    "    n_genes=40,\n",
    "    gene_names=10,\n",
    "    manual_added_genes= None, \n",
    "    heatmap_order_col='rho_axis',\n",
    "    groupby_replicate=True,\n",
    "    replicate_col=\"brain_region\",\n",
    "    replicate_palette=None,\n",
    "    replicate_order=None,\n",
    "    title=\"\",\n",
    "    filename=None, \n",
    "    out_path=None,\n",
    "    ylabel=\"Genes\",\n",
    "    xlabel=\"Cells\",\n",
    "    rna_cmap=\"Reds\",\n",
    "    mch_cmap=\"parula\",\n",
    "    rasterized=True,\n",
    "    show=True, \n",
    "    save=False,\n",
    "    image_path=None,\n",
    "    meth_min_max_quantiles=(0.02, 0.98),\n",
    "    rna_min_max_quantiles=(0.02, 0.98),\n",
    "    vmin=0, vmax=1,\n",
    "    color_splits=None,\n",
    "    figsize=(8,4),\n",
    "    title_fontsize=14,\n",
    "    def_fontsize=12, \n",
    "):\n",
    "    \"\"\"Plot expression heatmap for given genes and cell type\"\"\"\n",
    "    import PyComplexHeatmap as pch\n",
    "    from matplotlib.colors import TwoSlopeNorm\n",
    "\n",
    "    if filename is None: \n",
    "        filename = f\"joint_expr_heatmap_{title.replace(' ', '_')}\"\n",
    "    if replicate_palette is None:\n",
    "        replicate_palette = adata_rna.uns.get(f\"{replicate_col}_palette\", None)\n",
    "\n",
    "    df_col = adata_rna.obs[['Group', 'MS_NORM', 'MS_compartment']].copy()\n",
    "    if df_col.shape[0] > 50000:\n",
    "        df_col = df_col.sample(50000, random_state=42)\n",
    "    if groupby_replicate:\n",
    "        df_col[replicate_col] = adata_rna.obs[replicate_col]\n",
    "        df_col = df_col.sort_values([replicate_col, 'MS_NORM'])\n",
    "    else:\n",
    "        df_col = df_col.sort_values('MS_NORM')\n",
    "\n",
    "    df_row = df_genes.iloc[:n_genes].copy()\n",
    "    if manual_added_genes is not None:\n",
    "        df_manual = df_genes[ df_genes['gene'].isin(manual_added_genes) ].copy()\n",
    "        df_row = pd.concat([df_row, df_manual], axis=0)\n",
    "    df_row.drop_duplicates(subset='gene', keep='first', inplace=True)\n",
    "    if \"gene\" in df_row.columns:\n",
    "        df_row = df_row.set_index('gene')\n",
    "    toplot = df_row.index[:gene_names]\n",
    "    if manual_added_genes is not None:\n",
    "        toplot = list(toplot) + manual_added_genes\n",
    "    df_row['annot'] = [c if c in toplot else np.nan for c in df_row.index]\n",
    "    df_row = df_row.sort_values(heatmap_order_col, ascending=True)\n",
    "    # if \"gene\" in df_row.columns:\n",
    "    #     df_row = df_row.set_index('gene')\n",
    "\n",
    "    if hasattr(adata_rna.X, \"toarray\"):\n",
    "        df_expr_rna = adata_rna.X.toarray()\n",
    "    else: \n",
    "        df_expr_rna = adata_rna.X\n",
    "    df_expr_rna = pd.DataFrame(df_expr_rna, index=adata_rna.obs_names, columns=adata_rna.var_names).T\n",
    "    df_expr_rna = df_expr_rna.loc[df_row.index, df_col.index]\n",
    "    # df_expr_norm_rna = (\n",
    "    #     df_expr_rna\n",
    "    #     .subtract(df_expr_rna.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    #     .div(df_expr_rna.quantile(min_max_quantiles[1], axis=1) - df_expr_rna.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    # )\n",
    "    repl = []\n",
    "    for cc, df in df_col.groupby(replicate_col, observed=True): \n",
    "        df_samp = df_expr_rna[ df.index ].copy()\n",
    "        df_norm = (\n",
    "            df_samp\n",
    "            .subtract(df_samp.quantile(rna_min_max_quantiles[0], axis=1), axis=0)\n",
    "            .div(df_samp.quantile(rna_min_max_quantiles[1], axis=1) - df_samp.quantile(rna_min_max_quantiles[0], axis=1), axis=0)\n",
    "        )\n",
    "        repl.append(df_norm)\n",
    "    df_expr_norm_rna =  pd.concat(repl, axis=1)\n",
    "\n",
    "    if hasattr(adata_mch.X, \"toarray\"):\n",
    "        df_expr_mch = adata_mch.X.toarray()\n",
    "    else: \n",
    "        df_expr_mch = adata_mch.X\n",
    "    df_expr_mch = pd.DataFrame(df_expr_mch, index=adata_mch.obs_names, columns=adata_mch.var_names).T\n",
    "    df_expr_mch = df_expr_mch.loc[df_row.index, df_col.index]\n",
    "    # df_expr_norm_mch = (\n",
    "    #     df_expr_mch\n",
    "    #     .subtract(df_expr_mch.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    #     .div(df_expr_mch.quantile(min_max_quantiles[1], axis=1) - df_expr_mch.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    # )\n",
    "    repl = []\n",
    "    for cc, df in df_col.groupby(replicate_col, observed=True): \n",
    "        df_samp = df_expr_mch[ df.index ].copy()\n",
    "        df_norm = (\n",
    "            df_samp\n",
    "            .subtract(df_samp.quantile(meth_min_max_quantiles[0], axis=1), axis=0)\n",
    "            .div(df_samp.quantile(meth_min_max_quantiles[1], axis=1) - df_samp.quantile(meth_min_max_quantiles[0], axis=1), axis=0)\n",
    "        )\n",
    "        repl.append(df_norm)\n",
    "    df_expr_norm_mch =  pd.concat(repl, axis=1)\n",
    "\n",
    "    ms_score_norm = TwoSlopeNorm(vmin=-1, vcenter=0, vmax=1)\n",
    "    col_ha = pch.HeatmapAnnotation(\n",
    "        Brain_Region=pch.anno_simple(df_col[replicate_col], colors=replicate_palette, add_text=True, legend=False, height=3, text_kws={\"fontsize\":8, \"fontweight\":\"bold\"}, label=\"Brain Region\"),\n",
    "        Compartment=pch.anno_simple(df_col['MS_compartment'], colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, height=3, label=\"Compartment\"),\n",
    "        MatStr_Score=pch.anno_simple(df_col['MS_NORM'], cmap=\"coolwarm_r\", norm=ms_score_norm, height=3, label=\"MatStr Score\"),\n",
    "        verbose=1, axis=1, plot_legend=True, legend_gap=5, label_kws={'visible':False}\n",
    "    )\n",
    "\n",
    "    left_ha = pch.HeatmapAnnotation(\n",
    "        label=pch.anno_label(\n",
    "            df_row['annot'], merge=True, rotation=0, extend=True,\n",
    "            colors=\"black\", relpos=(1, 0.7), height=3, fontsize=def_fontsize,\n",
    "            arrowprops = dict(visible=True, linewidth=1, connectionstyle='arc3'),\n",
    "        ),\n",
    "        verbose=1, axis=0, orientation='left'\n",
    "    )\n",
    "\n",
    "    cm1 = pch.ClusterMapPlotter(\n",
    "        data=df_expr_norm_rna,\n",
    "        top_annotation=col_ha,\n",
    "        left_annotation=left_ha,\n",
    "        row_cluster=False,\n",
    "        col_cluster=False,\n",
    "        col_split=df_col[replicate_col] if groupby_replicate else None,\n",
    "        col_split_order=replicate_order,\n",
    "        row_dendrogram=False,\n",
    "        label=\"Expression\",\n",
    "        cmap=rna_cmap,\n",
    "        rasterized=rasterized,\n",
    "        ylabel=ylabel,\n",
    "        xlabel=\"RNA\",\n",
    "        xlabel_side=\"top\",\n",
    "        plot=False,\n",
    "        vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "\n",
    "    col_ha = pch.HeatmapAnnotation(\n",
    "        Brain_Region=pch.anno_simple(df_col[replicate_col], colors=replicate_palette, add_text=True, legend=False, height=3, text_kws={\"fontsize\":8, \"fontweight\":\"bold\"}, label=\"Brain Region\"),\n",
    "        Compartment=pch.anno_simple(df_col['MS_compartment'], colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, height=3, label=\"Compartment\"),\n",
    "        MatStr_Score=pch.anno_simple(df_col['MS_NORM'], cmap=\"coolwarm_r\", norm=ms_score_norm, height=3, label=\"MatStr Score\"),\n",
    "        verbose=1, axis=1, plot_legend=True, legend_gap=5,\n",
    "    )\n",
    "\n",
    "    cm2 = pch.ClusterMapPlotter(\n",
    "        data=df_expr_norm_mch,\n",
    "        top_annotation=col_ha,\n",
    "        # left_annotation=left_ha,\n",
    "        row_cluster=False,\n",
    "        col_cluster=False,\n",
    "        col_split=df_col[replicate_col] if groupby_replicate else None,\n",
    "        col_split_order=replicate_order,\n",
    "        row_dendrogram=False,\n",
    "        col_dendrogram=False,\n",
    "        label=\"Score\",\n",
    "        cmap=mch_cmap,\n",
    "        rasterized=rasterized,\n",
    "        ylabel=ylabel,\n",
    "        xlabel=\"mCH\",\n",
    "        xlabel_side=\"top\",\n",
    "        plot=False,\n",
    "        vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "\n",
    "\n",
    "    plt.figure(figsize=figsize)\n",
    "    ax, legend_axes = pch.composite(cmlist=[cm1, cm2], main=0, legend_hpad=3, col_gap=0.1)\n",
    "    for _leg in legend_axes: \n",
    "        _leg.set_rasterized(False)\n",
    "    ax.set_title(title, fontsize=title_fontsize, rasterized=False)\n",
    "\n",
    "    # plot custom spines # For coloring the region in between the col splits as something other than white! \n",
    "    if color_splits is not None: \n",
    "        for cm in [cm1, cm2]:\n",
    "            for i in range(cm.heatmap_axes.shape[0]):\n",
    "                for j in range(cm.heatmap_axes.shape[1]-1):\n",
    "                    # if i != j:\n",
    "                    #     continue\n",
    "                    ax = cm.heatmap_axes[i][j]\n",
    "                    for side in [\"right\"]:\n",
    "                        ax.spines[side].set_visible(True)\n",
    "                        ax.spines[side].set_color(color_splits)\n",
    "                        ax.spines[side].set_linewidth(2)\n",
    "\n",
    "    if save and image_path is not None: \n",
    "        print(\"saving\")\n",
    "        plt.savefig(image_path / f\"{filename}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "        plt.savefig(image_path / f\"{filename}.pdf\", dpi=300, bbox_inches=\"tight\", transparent=False)\n",
    "        # plt.savefig(image_path / f\"{filename}.svg\", dpi=300, bbox_inches=\"tight\")\n",
    "        # pass\n",
    "    if show: \n",
    "        plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8aca578",
   "metadata": {},
   "source": [
    "# SparkX Summary Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b0a1cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "level = \"Subclass\"\n",
    "_let = level.lower()[0]\n",
    "res_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/CPS/sparkx_br{_let}/sparkx_per_replicate_results.csv\"\n",
    "meta_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/CPS/sparkx_br{_let}/sparkx_meta_results.csv\"\n",
    "\n",
    "df_meta = pd.read_csv(meta_path)\n",
    "sparkx_df = pd.read_csv(res_path)\n",
    "\n",
    "common_genes = adata.var_names.intersection(spatial_mch.var_names)\n",
    "df_meta = df_meta[df_meta['gene'].isin(common_genes)].copy()\n",
    "sparkx_df = sparkx_df[sparkx_df['gene'].isin(common_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08433090",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do not want results from CAT\n",
    "sparkx_df = sparkx_df[sparkx_df['replicate'] != \"CAT\"].copy()\n",
    "sparkx_df['direction'] = (sparkx_df['rho_axis'] < 0).map({True: 'Striosome', False: 'Matrix'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05f9907e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sig_counts from meta analysis\n",
    "sig_counts = df_meta.loc[df_meta['fdr'] < 1e-2].groupby(\"cell_type\")['direction'].value_counts().reset_index().rename(columns={'count' : 'n_sig'})\n",
    "sig_counts['direction'] = sig_counts['direction'].map({'down': 'Striosome', 'up':'Matrix'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907cfa75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Compute replicate-level counts ---\n",
    "fdr_col = \"p_sparkx\"\n",
    "fdr_thresh = 1e-2\n",
    "\n",
    "# sig_counts = (\n",
    "#     sparkx_df.groupby([\"cell_type\", \"replicate\", \"direction\"], observed=True)\n",
    "#     .apply(lambda x: ((x[fdr_col] < fdr_thresh)).sum())\n",
    "#     .reset_index(name=\"n_sig\")\n",
    "# )\n",
    "# sig_counts = sig_counts.reset_index(drop=True)\n",
    "# sig_counts = sig_counts[sig_counts['cell_type'] != \"unknown\"]\n",
    "\n",
    "# cell_types = sig_counts[\"cell_type\"].unique()\n",
    "# directions = sig_counts[\"direction\"].unique()\n",
    "\n",
    "# Updated mid-soft palette\n",
    "palette = {\n",
    "    \"Matrix\":    (0.30, 0.50, 0.90),\n",
    "    \"Striosome\": (0.90, 0.45, 0.45),\n",
    "}\n",
    "\n",
    "\n",
    "# --- Set up figure ---\n",
    "fig, ax = plt.subplots(figsize=(8, 3))\n",
    "\n",
    "x = np.arange(len(cell_types))\n",
    "n_dir = len(directions)\n",
    "width = 0.8 / n_dir\n",
    "\n",
    "jitter_scale = 0.3   # jitter occupies 30% of the bar width\n",
    "rng = np.random.default_rng(42)\n",
    "\n",
    "# --- Plot bars, error bars, and jittered points ---\n",
    "for i, direction in enumerate(directions):\n",
    "\n",
    "    # x-pos of this direction’s bars\n",
    "    x_pos = x - 0.4 + i * width + width / 2\n",
    "\n",
    "    means = []\n",
    "    errs = []\n",
    "    points_per_ct = []   # ensures exact alignment with x_pos\n",
    "\n",
    "    for ct in cell_types:\n",
    "        vals = sig_counts.loc[\n",
    "            (sig_counts[\"cell_type\"] == ct) &\n",
    "            (sig_counts[\"direction\"] == direction),\n",
    "            \"n_sig\"\n",
    "        ].values\n",
    "        \n",
    "        means.append(vals.mean())\n",
    "        errs.append(vals.std(ddof=1))\n",
    "        points_per_ct.append(vals)  # aligned list → FIX\n",
    "\n",
    "    # --- Bars ---\n",
    "    ax.bar(\n",
    "        x_pos, means, width=width,\n",
    "        color=palette.get(direction, \"gray\"),\n",
    "        edgecolor=\"black\", linewidth=0.5,\n",
    "        label=direction,\n",
    "        rasterized=RASTERIZED, zorder=2\n",
    "    )\n",
    "\n",
    "    # --- Error bars ---\n",
    "    ax.errorbar(\n",
    "        x_pos, means, yerr=errs,\n",
    "        fmt=\"none\", ecolor=\"black\",\n",
    "        elinewidth=1, capsize=3,\n",
    "        zorder=3\n",
    "    )\n",
    "\n",
    "    # --- Jittered points ---\n",
    "    for xi, vals in zip(x_pos, points_per_ct):\n",
    "        if len(vals) == 0:\n",
    "            continue\n",
    "        jitter = rng.uniform(\n",
    "            -width * jitter_scale,\n",
    "            width * jitter_scale,\n",
    "            size=len(vals)\n",
    "        )\n",
    "        ax.scatter(\n",
    "            xi + jitter, vals,\n",
    "            color=\"black\", s=2,\n",
    "            zorder=4,\n",
    "            rasterized=RASTERIZED\n",
    "        )\n",
    "\n",
    "# --- Axes styling ---\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(cell_types, rotation=30, ha=\"right\", fontsize=8, rasterized=RASTERIZED)\n",
    "\n",
    "ax.set_ylabel(f\"# Significant Genes (FDR<{fdr_thresh})\", rasterized=RASTERIZED)\n",
    "# ax.set_xlabel(\"Cell Type\", rasterized=RASTERIZED)\n",
    "ax.set_title(\"Significant Compartment Enriched Genes per Cell Type\", rasterized=RASTERIZED, fontsize=14)\n",
    "\n",
    "# --- Color x-labels conditionally ---\n",
    "color_ct = [\n",
    "    ct for ct in cell_types\n",
    "    if sig_counts[sig_counts[\"cell_type\"] == ct][\"n_sig\"].sum() > 25\n",
    "]\n",
    "\n",
    "for ticklabel in ax.get_xticklabels():\n",
    "    ticklabel.set_color(\"black\" if ticklabel.get_text() in color_ct else \"lightgray\")\n",
    "\n",
    "ax.legend(frameon=False)\n",
    "\n",
    "fig.tight_layout()\n",
    "# fig.savefig(image_path / f\"num_sig_genes_per_celltype_full_{level}.png\",dpi=300, bbox_inches=\"tight\")\n",
    "# fig.savefig(image_path / f\"num_sig_genes_per_celltype_full_{level}.pdf\",dpi=300, bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48fe3aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fdr_col = \"p_sparkx\"\n",
    "# fdr_thresh = 1e-2\n",
    "# sig_counts = (\n",
    "#     sparkx_df.groupby([\"cell_type\", \"replicate\", \"direction\"], observed=True)\n",
    "#     .apply(lambda x: ((x[fdr_col] < fdr_thresh)).sum())\n",
    "#     .reset_index(name=\"n_sig\")\n",
    "# )\n",
    "# sig_counts = sig_counts.reset_index()\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(8,3))\n",
    "# sns.barplot(\n",
    "#     data=sig_counts, x=\"cell_type\", y=\"n_sig\", hue=\"direction\",\n",
    "#     palette = {\"Matrix\": \"blue\", \"Striosome\": \"red\"},\n",
    "#     errorbar=None, err_kws={\"color\": \"black\", \"linewidth\": 1}, capsize=0.2,\n",
    "#     ax=ax, rasterized=RASTERIZED\n",
    "# )\n",
    "# ax.set_ylabel(f\"# Significant Genes (FDR<{fdr_thresh})\", rasterized=RASTERIZED)\n",
    "# ax.set_xlabel(\"Cell Type\", rasterized=RASTERIZED)\n",
    "# ax.set_title(\"Significant Compartment Enriched Genes per Cell Type\", rasterized=RASTERIZED)\n",
    "# ax.set_xticks(ax.get_xticks())\n",
    "# ax.set_xticklabels(ax.get_xticklabels(), rotation=30, ha=\"right\", rasterized=RASTERIZED)\n",
    "\n",
    "\n",
    "# color_ct = []\n",
    "# for _ct in sig_counts['cell_type'].unique():\n",
    "#     if sig_counts[(sig_counts['cell_type'] == _ct)]['n_sig'].sum() > 5: \n",
    "#         color_ct.append(_ct)\n",
    "\n",
    "# for i, ticklabel in enumerate(ax.get_xticklabels()): \n",
    "#     ticklabel.set_color('black' if ticklabel.get_text() in color_ct else 'lightgray')\n",
    "\n",
    "# plt.tight_layout()\n",
    "# plt.savefig(image_path / f\"num_sig_genes_per_celltype_full_{level}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "# plt.savefig(image_path / f\"num_sig_genes_per_celltype_full_{level}.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "# plt.show()\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437a6335",
   "metadata": {},
   "outputs": [],
   "source": [
    "fdr_col = \"p_sparkx\"\n",
    "fdr_thresh = 1e-5\n",
    "sig_counts = (\n",
    "    sparkx_df.groupby([\"cell_type\", \"replicate\"], observed=True)\n",
    "    .apply(lambda x: ((x[fdr_col] < fdr_thresh)).sum())\n",
    "    .reset_index(name=\"n_sig\")\n",
    "    .groupby([\"cell_type\"], observed=True)\n",
    "    .agg({'n_sig': ['mean', 'std']})\n",
    ")\n",
    "sig_counts.columns = sig_counts.columns.droplevel(0)\n",
    "sig_counts = (\n",
    "    sig_counts.reset_index()\n",
    "    .query(\"mean > 1\")\n",
    "    .sort_values('mean', ascending=True)\n",
    ")\n",
    "sig_counts['color'] = sig_counts['cell_type'].map(adata.uns[f\"{level}_palette\"])\n",
    "sig_counts = sig_counts[sig_counts['cell_type'] != \"unknown\"]\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4,3))\n",
    "ax.barh(\n",
    "    data=sig_counts, y=\"cell_type\", width=\"mean\", color=sig_counts['color'],\n",
    "    rasterized=RASTERIZED\n",
    ")\n",
    "ax.set_ylabel(\"Cell Type\", rasterized=RASTERIZED)\n",
    "ax.set_xlabel(f\"# Significant Genes (FDR<{fdr_thresh})\", rasterized=RASTERIZED)\n",
    "ax.set_title(\"Significant Compartment Enriched Genes per Cell Type\", rasterized=RASTERIZED, x=0.2)\n",
    "\n",
    "group_to_nt = adata.obs[[level, \"Neighborhood\"]].drop_duplicates().set_index(level).to_dict()['Neighborhood']\n",
    "tick_to_nn_pal = {tick: adata.uns['m3c_neighborhood_palette'].get(nt, '#808080') for tick, nt in group_to_nt.items()}\n",
    "for i, ticklabel in enumerate(ax.get_yticklabels()): \n",
    "    ticklabel.set_color(tick_to_nn_pal.get(ticklabel.get_text(), '#808080'))\n",
    "\n",
    "plt.tight_layout()\n",
    "# plt.savefig(image_path / f\"num_sig_genes_per_celltype_{level}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "# plt.savefig(image_path / f\"num_sig_genes_per_celltype_{level}.pdf\", dpi=300, bbox_inches=\"tight\", pad_inches=0.5)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d54933b5",
   "metadata": {},
   "source": [
    "## Trial: Overlap with Wubin DMGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2846d5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0caa91db",
   "metadata": {},
   "outputs": [],
   "source": [
    "dmg_res = pd.read_csv(\"/anvil/projects/x-mcb130189/Wubin/BG/DMG/merged_dmg.one_vs_rest.tsv\", sep=\"\\t\")\n",
    "msn_dmg = dmg_res.query(\"Group=='Group' & Parent=='MSN' & mc_type=='CH'\")\n",
    "# display(dmg_res.head())\n",
    "del dmg_res\n",
    "msn_dmg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30caae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_genes = adata.var_names\n",
    "common_genes = spatial_genes.intersection(msn_dmg['Gene'])\n",
    "msn_dmg = msn_dmg[ msn_dmg['Gene'].isin(common_genes) ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6968137",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix_dmgs = msn_dmg.loc[(msn_dmg['groupA'] == \"STRd D2 Matrix MSN\") | (msn_dmg['groupA'] == \"STRd D1 Matrix MSN\") ]\n",
    "striosome_dmgs = msn_dmg.loc[(msn_dmg['groupA'] == \"STRd D1 Striosome MSN\") | (msn_dmg['groupA'] == \"STRd D2 Striosome MSN\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb7adced",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ct = \"STR D2 MSN\"\n",
    "fdr_thresh = 1e-5\n",
    "df_sub = (\n",
    "    sparkx_df[sparkx_df['cell_type'] == _ct]\n",
    "    .query(f\"{fdr_col} < {fdr_thresh} & {fdr_col} >= 0\")\n",
    "    .groupby([\"gene\", \"replicate\"])[[fdr_col, 'rho_axis']]\n",
    "    .median()\n",
    "    .sort_values(fdr_col, ascending=True)\n",
    "    .reset_index()\n",
    "    .drop_duplicates(subset='gene', keep='first')\n",
    ")\n",
    "df_sub.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0e3da6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sub_str = df_sub.loc[df_sub['rho_axis'] < 0]\n",
    "df_sub_mat = df_sub.loc[df_sub['rho_axis'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5626852",
   "metadata": {},
   "outputs": [],
   "source": [
    "naming_map = {\"STR D1 MSN\" : [\"STRd D1 Matrix MSN\", \"STRd D1 Striosome MSN\"], \n",
    "              \"STR D2 MSN\" : [\"STRd D2 Matrix MSN\", \"STRd D2 Striosome MSN\"] }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "951e3c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "dmg_fdr_thresh = 1e-5\n",
    "sub_matrix_dmgs = matrix_dmgs[(matrix_dmgs[\"groupA\"].isin(naming_map[_ct])) & (matrix_dmgs['fdr_bh'] < dmg_fdr_thresh)].copy()\n",
    "sub_matrix_dmgs.drop_duplicates(subset='Gene', keep='first', inplace=True)\n",
    "sub_matrix_dmgs.shape\n",
    "\n",
    "mat_ab = df_sub_mat['gene'].isin(sub_matrix_dmgs['Gene']).sum()\n",
    "mat_a = df_sub_mat.shape[0] - mat_ab\n",
    "mat_b = sub_matrix_dmgs.shape[0] - mat_ab\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "venn2(\n",
    "    subsets= (mat_a, mat_b, mat_ab),\n",
    "    set_labels = ('Matrix Enriched Genes', 'Matrix DMGs'),\n",
    "    set_colors = (adata.uns[f\"{level}_palette\"]['STR D1 MSN'], 'blue'),\n",
    "    ax=ax)\n",
    "ax.set_title(\"Matrix Enriched Genes vs Matrix DMGs\\n(STR D1 MSNs)\", rasterized=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90e986d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dmg_fdr_thresh = 1e-5\n",
    "sub_str_dmgs = striosome_dmgs[(striosome_dmgs[\"groupA\"].isin(naming_map[_ct])) & (striosome_dmgs['fdr_bh'] < dmg_fdr_thresh)].copy()\n",
    "sub_str_dmgs.drop_duplicates(subset='Gene', keep='first', inplace=True)\n",
    "sub_str_dmgs.shape\n",
    "\n",
    "str_ab = df_sub_str['gene'].isin(sub_str_dmgs['Gene']).sum()\n",
    "str_a = df_sub_str.shape[0] - str_ab\n",
    "str_b = sub_str_dmgs.shape[0] - str_ab\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "venn2(\n",
    "    subsets= (str_a, str_b, str_ab),\n",
    "    set_labels = ('Striosome Enriched Genes', 'Striosome DMGs'),\n",
    "    set_colors = (adata.uns[f\"MSN_Groups_palette\"][naming_map[_ct][0]], 'blue'),\n",
    "    ax=ax)\n",
    "ax.set_title(\"Striosome Enriched Genes vs Striosome DMGs\\n(STR D1 MSNs)\", rasterized=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99ed7ba",
   "metadata": {},
   "source": [
    "# SPARKX Heatmaps\n",
    "Start from here to just plot the heatmaps after running the initial cells and the helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6db66e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "level = \"Subclass\"\n",
    "_let = level.lower()[0]\n",
    "res_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/CPS/sparkx_br{_let}/sparkx_per_replicate_results.csv\"\n",
    "meta_path = f\"/home/x-aklein2/projects/aklein/BICAN/BG/data/CPS/sparkx_br{_let}/sparkx_meta_results.csv\"\n",
    "\n",
    "df_meta = pd.read_csv(meta_path)\n",
    "sparkx_df = pd.read_csv(res_path)\n",
    "\n",
    "common_genes = adata.var_names.intersection(spatial_mch.var_names)\n",
    "df_meta = df_meta[df_meta['gene'].isin(common_genes)].copy()\n",
    "sparkx_df = sparkx_df[sparkx_df['gene'].isin(common_genes)].copy()\n",
    "\n",
    "# Do not want results from CAT\n",
    "sparkx_df = sparkx_df[sparkx_df['replicate'] != \"CAT\"].copy()\n",
    "sparkx_df['direction'] = (sparkx_df['rho_axis'] > 0).map({True: 'Striosome', False: 'Matrix'})\n",
    "\n",
    "# --- Compute replicate-level counts ---\n",
    "fdr_col = \"p_sparkx\"\n",
    "fdr_thresh = 1e-5\n",
    "\n",
    "group_to_nt = adata.obs[[level, \"Neighborhood\"]].drop_duplicates().set_index(level).to_dict()['Neighborhood']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defc764c",
   "metadata": {},
   "outputs": [],
   "source": [
    "transfer_cols = ['brain_region_corr', 'MS_NORM', 'MS_SCORE', 'MS_compartment']\n",
    "for col in transfer_cols:\n",
    "    spatial_mch.obs[col] = adata.obs.loc[spatial_mch.obs.index, col].copy()\n",
    "\n",
    "joint_cells = adata.obs_names.intersection(spatial_mch.obs_names)\n",
    "adata_joint_rna = adata[joint_cells].copy()\n",
    "adata_joint_mch = spatial_mch[joint_cells].copy()\n",
    "\n",
    "repr_order = [\"CaH\", \"CaB\", \"Pu\", \"NAC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8060d056",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture\n",
    "# # TODO: Add Manual Labeling of Genes\n",
    "# topn = 10\n",
    "# for _ct in sig_counts['cell_type'].unique():\n",
    "#     df_sub = (\n",
    "#         sparkx_df[sparkx_df['cell_type'] == _ct]\n",
    "#         .query(f\"{fdr_col} < {fdr_thresh} & {fdr_col} >= 0\")\n",
    "#         .groupby([\"gene\", \"replicate\"])[[fdr_col, 'rho_axis']]\n",
    "#         .median()\n",
    "#         .sort_values(fdr_col, ascending=True)\n",
    "#         .reset_index()\n",
    "#         # .drop_duplicates(subset='gene', keep='first')\n",
    "#     )\n",
    "#     gene_vc = df_sub['gene'].value_counts()\n",
    "#     ns = len(df_sub['gene'].unique())\n",
    "#     if ns < 50: \n",
    "#         gvc_thr = 1\n",
    "#     elif ns < 100: \n",
    "#         gvc_thr = 2\n",
    "#     else: \n",
    "#         gvc_thr = 3\n",
    "#     keep_genes = gene_vc[gene_vc >= gvc_thr].index\n",
    "#     df_sub = df_sub[df_sub['gene'].isin(keep_genes)].copy()\n",
    "#     df_sub = df_sub.drop_duplicates(subset='gene', keep='first')\n",
    "\n",
    "#     nt = group_to_nt.get(_ct)\n",
    "#     ns = len(df_sub['gene'].unique())\n",
    "#     print(_ct, ns, nt)\n",
    "#     _ct_filename = _ct.replace(\"/\", \"_\").replace(\" \", \"_\")\n",
    "#     if ns < 5: \n",
    "#         continue\n",
    "\n",
    "#     if nt in ['Nonneuron', \"unknown\"]: \n",
    "#         # Plot only RNA\n",
    "#         # pass\n",
    "#         adata_ct = adata[(adata.obs[level] == _ct) & (~adata.obs['MS_NORM'].isna())].copy()\n",
    "#         adata_ct = adata_ct[adata_ct.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "#         adata_ct = _downsample_reference(adata_ct, cluster_col=\"MS_compartment\", max_cluster_size=6000, min_cluster_size=0)\n",
    "#         adata_ct.obs['brain_region_corr'] = pd.Categorical(adata_ct.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "        \n",
    "#         rep_palette = {k : v for k, v in adata_ct.uns.get(\"brain_region_corr_palette\", {}).items() if k in adata_ct.obs['brain_region_corr'].unique()}\n",
    "#         expr_heatmap(adata_ct, df_sub, n_genes=ns, gene_names=10, ylabel=None, color_splits=\"white\",\n",
    "#                     rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "#                     replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "#                     cmap='agSunset', vmin=0, vmax=1, min_max_quantiles=(0.25, 0.99), \n",
    "#                     filename=f\"sparkx_heatmap_{_ct_filename}\",\n",
    "#         )\n",
    "#         # break\n",
    "#     else: \n",
    "#         # Plot joint RNA + mCH\n",
    "#         adata_rna = adata_joint_rna[(adata_joint_rna.obs[level] == _ct) & (~adata_joint_rna.obs['MS_NORM'].isna())].copy()\n",
    "\n",
    "#         adata_rna = adata_rna[adata_rna.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "#         adata_rna = _downsample_reference(adata_rna, cluster_col=\"MS_compartment\", max_cluster_size=4000, min_cluster_size=0)\n",
    "#         adata_rna.obs['brain_region_corr'] = pd.Categorical(adata_rna.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "#         adata_mch = adata_joint_mch[(adata_joint_mch.obs[level] == _ct) & (~adata_joint_mch.obs['MS_NORM'].isna())].copy()\n",
    "#         adata_mch = adata_mch[adata_mch.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "#         adata_mch.obs['brain_region_corr'] = pd.Categorical(adata_mch.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "        \n",
    "#         rep_palette = {k : adata_rna.uns[\"brain_region_corr_palette\"].get(k, None) for k in repr_order}\n",
    "#         joint_expr_heatmap(adata_rna, adata_mch, df_sub, n_genes=ns, gene_names=topn, ylabel=None, color_splits=\"white\", \n",
    "#                     rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "#                     replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "#                     rna_cmap='agSunset', vmin=0, vmax=1, meth_min_max_quantiles=(0.1, 0.9), rna_min_max_quantiles=(0.25, 0.99),\n",
    "#                     figsize=(12,6), filename=f\"sparkx_heatmap_{_ct_filename}\",\n",
    "#         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5632aaa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture\n",
    "manual_genes = ['BACH2', 'KIRREL3', 'GRM1', 'PDYN']\n",
    "topn = 5\n",
    "# max_genes = 40\n",
    "_ct = \"STR D1 MSN\"\n",
    "df_sub = (\n",
    "    sparkx_df[sparkx_df['cell_type'] == _ct]\n",
    "    .query(f\"{fdr_col} < {fdr_thresh} & {fdr_col} >= 0 | gene in {manual_genes}\")\n",
    "    .groupby([\"gene\", \"replicate\"])[[fdr_col, 'rho_axis']]\n",
    "    .median()\n",
    "    .sort_values(fdr_col, ascending=True)\n",
    "    .reset_index()\n",
    "    # .drop_duplicates(subset='gene', keep='first')\n",
    ")\n",
    "gene_vc = df_sub['gene'].value_counts()\n",
    "ns = len(df_sub['gene'].unique())\n",
    "if ns < 50: \n",
    "    gvc_thr = 1\n",
    "elif ns < 100: \n",
    "    gvc_thr = 2\n",
    "else: \n",
    "    gvc_thr = 3\n",
    "keep_genes = gene_vc[gene_vc >= gvc_thr].index\n",
    "df_sub = df_sub[df_sub['gene'].isin(keep_genes)].copy()\n",
    "df_sub = df_sub.drop_duplicates(subset='gene', keep='first')\n",
    "# df_sub = df_sub.head(max_genes)\n",
    "\n",
    "nt = group_to_nt.get(_ct)\n",
    "ns = len(df_sub['gene'].unique())\n",
    "print(_ct, ns, nt)\n",
    "_ct_filename = _ct.replace(\"/\", \"_\").replace(\" \", \"_\")\n",
    "# if ns < 5: \n",
    "#     continue\n",
    "\n",
    "if nt in ['Nonneuron', \"unknown\"]: \n",
    "    # Plot only RNA\n",
    "    # pass\n",
    "    adata_ct = adata[(adata.obs[level] == _ct) & (~adata.obs['MS_NORM'].isna())].copy()\n",
    "    adata_ct = adata_ct[adata_ct.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_ct = _downsample_reference(adata_ct, cluster_col=\"MS_compartment\", max_cluster_size=4000, min_cluster_size=0)\n",
    "    adata_ct.obs['brain_region_corr'] = pd.Categorical(adata_ct.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "    \n",
    "    rep_palette = {k : v for k, v in adata_ct.uns.get(\"brain_region_corr_palette\", {}).items() if k in adata_ct.obs['brain_region_corr'].unique()}\n",
    "    expr_heatmap(adata_ct, df_sub, n_genes=ns, gene_names=10, ylabel=None, color_splits=\"white\",\n",
    "                rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "                replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "                cmap='agSunset', vmin=0, vmax=1, min_max_quantiles=(0.25, 0.99), \n",
    "                filename=f\"sparkx_heatmap_{_ct_filename}\",\n",
    "    )\n",
    "    # break\n",
    "else: \n",
    "    # Plot joint RNA + mCH\n",
    "    adata_rna = adata_joint_rna[(adata_joint_rna.obs[level] == _ct) & (~adata_joint_rna.obs['MS_NORM'].isna())].copy()\n",
    "\n",
    "    adata_rna = adata_rna[adata_rna.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_rna = _downsample_reference(adata_rna, cluster_col=\"MS_compartment\", max_cluster_size=8000, min_cluster_size=0)\n",
    "    adata_rna.obs['brain_region_corr'] = pd.Categorical(adata_rna.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "    adata_mch = adata_joint_mch[(adata_joint_mch.obs[level] == _ct) & (~adata_joint_mch.obs['MS_NORM'].isna())].copy()\n",
    "    adata_mch = adata_mch[adata_mch.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_mch.obs['brain_region_corr'] = pd.Categorical(adata_mch.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "    \n",
    "    rep_palette = {k : adata_rna.uns[\"brain_region_corr_palette\"].get(k, None) for k in repr_order}\n",
    "    joint_expr_heatmap(adata_rna, adata_mch, df_sub, n_genes=ns, gene_names=topn, manual_added_genes = manual_genes, \n",
    "                ylabel=None, color_splits=\"white\", \n",
    "                rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "                replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "                rna_cmap='agSunset', vmin=0, vmax=1, meth_min_max_quantiles=(0.1, 0.9), rna_min_max_quantiles=(0.25, 0.99),\n",
    "                figsize=(12,6), filename=f\"sparkx_heatmap_{_ct_filename}_2\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d7288e",
   "metadata": {},
   "outputs": [],
   "source": [
    "manual_genes = ['RASGRF2', 'GLP1R', 'GALNT17', 'ALK']\n",
    "topn = 5\n",
    "_ct = \"CN ST18 GABA\"\n",
    "df_sub = (\n",
    "    sparkx_df[sparkx_df['cell_type'] == _ct]\n",
    "    .query(f\"{fdr_col} < {fdr_thresh} & {fdr_col} >= 0 | gene in {manual_genes}\")\n",
    "    .groupby([\"gene\", \"replicate\"])[[fdr_col, 'rho_axis']]\n",
    "    .median()\n",
    "    .sort_values(fdr_col, ascending=True)\n",
    "    .reset_index()\n",
    "    # .drop_duplicates(subset='gene', keep='first')\n",
    ")\n",
    "gene_vc = df_sub['gene'].value_counts()\n",
    "ns = len(df_sub['gene'].unique())\n",
    "if ns < 50: \n",
    "    gvc_thr = 1\n",
    "elif ns < 100: \n",
    "    gvc_thr = 2\n",
    "else: \n",
    "    gvc_thr = 3\n",
    "keep_genes = gene_vc[gene_vc >= gvc_thr].index\n",
    "df_sub = df_sub[df_sub['gene'].isin(keep_genes)].copy()\n",
    "df_sub = df_sub.drop_duplicates(subset='gene', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b7ea926",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sig_counts from meta analysis\n",
    "sig_counts = df_meta.loc[df_meta['fdr'] < 0.05].groupby(\"cell_type\")['direction'].value_counts().reset_index().rename(columns={'count' : 'n_sig'})\n",
    "sig_counts['direction'] = sig_counts['direction'].map({'down': 'Striosome', 'up':'Matrix'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bf946b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ct = \"STR D2 MSN\"\n",
    "fdr_thresh = 1e-5\n",
    "df_sub2 = (\n",
    "    df_meta.loc[df_meta['cell_type'] == _ct]\n",
    "    .query(f\"fdr < {fdr_thresh}\")\n",
    ")\n",
    "_ct_filename = _ct.replace(\"/\", \"_\").replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42722c14",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ct = adata[(adata.obs[level] == _ct) & (~adata.obs['MS_NORM'].isna())].copy()\n",
    "adata_ct = adata_ct[adata_ct.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "adata_ct = _downsample_reference(adata_ct, cluster_col=\"MS_compartment\", max_cluster_size=4000, min_cluster_size=0)\n",
    "adata_ct.obs['brain_region_corr'] = pd.Categorical(adata_ct.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "rep_palette = {k : v for k, v in adata_ct.uns.get(\"brain_region_corr_palette\", {}).items() if k in adata_ct.obs['brain_region_corr'].unique()}\n",
    "expr_heatmap(adata_ct, df_sub2, n_genes=ns, gene_names=10, ylabel=None, color_splits=\"white\", heatmap_order_col=\"meta_Z\",\n",
    "            rasterized=True, show=True, save=False, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "            replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "            cmap='agSunset', vmin=0, vmax=1, min_max_quantiles=(0.25, 0.99), \n",
    "            filename=f\"sparkx_heatmap_{_ct_filename}\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44f07d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_rna = adata_joint_rna[(adata_joint_rna.obs[level] == _ct) & (~adata_joint_rna.obs['MS_NORM'].isna())].copy()\n",
    "\n",
    "adata_rna = adata_rna[adata_rna.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "adata_rna = _downsample_reference(adata_rna, cluster_col=\"MS_compartment\", max_cluster_size=2000, min_cluster_size=0)\n",
    "adata_rna.obs['brain_region_corr'] = pd.Categorical(adata_rna.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "adata_mch = adata_joint_mch[(adata_joint_mch.obs[level] == _ct) & (~adata_joint_mch.obs['MS_NORM'].isna())].copy()\n",
    "adata_mch = adata_mch[adata_mch.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "adata_mch.obs['brain_region_corr'] = pd.Categorical(adata_mch.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "rep_palette = {k : adata_rna.uns[\"brain_region_corr_palette\"].get(k, None) for k in repr_order}\n",
    "joint_expr_heatmap(adata_rna, adata_mch, df_sub2, n_genes=ns, gene_names=topn, manual_added_genes = manual_genes, \n",
    "            ylabel=None, color_splits=\"white\", heatmap_order_col=\"meta_Z\",\n",
    "            rasterized=True, show=True, save=False, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "            replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "            rna_cmap='agSunset', vmin=0, vmax=1, meth_min_max_quantiles=(0.1, 0.9), rna_min_max_quantiles=(0.25, 0.99),\n",
    "            figsize=(12,6), filename=f\"sparkx_heatmap_{_ct_filename}_2\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6204409",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture\n",
    "# TODO: Add Manual Labeling of Genes\n",
    "manual_genes = ['RASGRF2', 'GLP1R', 'GALNT17', 'ALK']\n",
    "topn = 5\n",
    "_ct = \"CN ST18 GABA\"\n",
    "df_sub = (\n",
    "    sparkx_df[sparkx_df['cell_type'] == _ct]\n",
    "    .query(f\"{fdr_col} < {fdr_thresh} & {fdr_col} >= 0 | gene in {manual_genes}\")\n",
    "    .groupby([\"gene\", \"replicate\"])[[fdr_col, 'rho_axis']]\n",
    "    .median()\n",
    "    .sort_values(fdr_col, ascending=True)\n",
    "    .reset_index()\n",
    "    # .drop_duplicates(subset='gene', keep='first')\n",
    ")\n",
    "gene_vc = df_sub['gene'].value_counts()\n",
    "ns = len(df_sub['gene'].unique())\n",
    "if ns < 50: \n",
    "    gvc_thr = 1\n",
    "elif ns < 100: \n",
    "    gvc_thr = 2\n",
    "else: \n",
    "    gvc_thr = 3\n",
    "keep_genes = gene_vc[gene_vc >= gvc_thr].index\n",
    "df_sub = df_sub[df_sub['gene'].isin(keep_genes)].copy()\n",
    "df_sub = df_sub.drop_duplicates(subset='gene', keep='first')\n",
    "\n",
    "nt = group_to_nt.get(_ct)\n",
    "ns = len(df_sub['gene'].unique())\n",
    "print(_ct, ns, nt)\n",
    "_ct_filename = _ct.replace(\"/\", \"_\").replace(\" \", \"_\")\n",
    "# if ns < 5: \n",
    "#     continue\n",
    "\n",
    "if nt in ['Nonneuron', \"unknown\"]: \n",
    "    # Plot only RNA\n",
    "    # pass\n",
    "    adata_ct = adata[(adata.obs[level] == _ct) & (~adata.obs['MS_NORM'].isna())].copy()\n",
    "    adata_ct = adata_ct[adata_ct.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_ct = _downsample_reference(adata_ct, cluster_col=\"MS_compartment\", max_cluster_size=4000, min_cluster_size=0)\n",
    "    adata_ct.obs['brain_region_corr'] = pd.Categorical(adata_ct.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "    \n",
    "    rep_palette = {k : v for k, v in adata_ct.uns.get(\"brain_region_corr_palette\", {}).items() if k in adata_ct.obs['brain_region_corr'].unique()}\n",
    "    expr_heatmap(adata_ct, df_sub, n_genes=ns, gene_names=10, ylabel=None, color_splits=\"white\",\n",
    "                rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "                replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "                cmap='agSunset', vmin=0, vmax=1, min_max_quantiles=(0.25, 0.99), \n",
    "                filename=f\"sparkx_heatmap_{_ct_filename}\",\n",
    "    )\n",
    "    # break\n",
    "else: \n",
    "    # Plot joint RNA + mCH\n",
    "    adata_rna = adata_joint_rna[(adata_joint_rna.obs[level] == _ct) & (~adata_joint_rna.obs['MS_NORM'].isna())].copy()\n",
    "\n",
    "    adata_rna = adata_rna[adata_rna.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_rna = _downsample_reference(adata_rna, cluster_col=\"MS_compartment\", max_cluster_size=2000, min_cluster_size=0)\n",
    "    adata_rna.obs['brain_region_corr'] = pd.Categorical(adata_rna.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "\n",
    "    adata_mch = adata_joint_mch[(adata_joint_mch.obs[level] == _ct) & (~adata_joint_mch.obs['MS_NORM'].isna())].copy()\n",
    "    adata_mch = adata_mch[adata_mch.obs['brain_region_corr'].isin(repr_order)].copy()\n",
    "    adata_mch.obs['brain_region_corr'] = pd.Categorical(adata_mch.obs['brain_region_corr'], categories=repr_order, ordered=True)\n",
    "    \n",
    "    rep_palette = {k : adata_rna.uns[\"brain_region_corr_palette\"].get(k, None) for k in repr_order}\n",
    "    joint_expr_heatmap(adata_rna, adata_mch, df_sub, n_genes=ns, gene_names=topn, manual_added_genes = manual_genes, \n",
    "                ylabel=None, color_splits=\"white\", \n",
    "                rasterized=True, show=False, save=True, image_path=image_path, title=f\"Top SPARK-X genes in {_ct}\",\n",
    "                replicate_col=\"brain_region_corr\", replicate_palette=rep_palette, replicate_order=repr_order,\n",
    "                rna_cmap='agSunset', vmin=0, vmax=1, meth_min_max_quantiles=(0.1, 0.9), rna_min_max_quantiles=(0.25, 0.99),\n",
    "                figsize=(12,6), filename=f\"sparkx_heatmap_{_ct_filename}_2\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e94598c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def joint_expr_heatmap(\n",
    "    adata_rna,\n",
    "    adata_mch,\n",
    "    df_genes,\n",
    "    n_genes=40,\n",
    "    gene_names=10,\n",
    "    manual_added_genes= None, \n",
    "    heatmap_order_col='rho_axis',\n",
    "    groupby_replicate=True,\n",
    "    replicate_col=\"brain_region\",\n",
    "    replicate_palette=None,\n",
    "    replicate_order=None,\n",
    "    title=\"\",\n",
    "    filename=None, \n",
    "    out_path=None,\n",
    "    ylabel=\"Genes\",\n",
    "    xlabel=\"Cells\",\n",
    "    rna_cmap=\"Reds\",\n",
    "    mch_cmap=\"parula\",\n",
    "    rasterized=True,\n",
    "    show=True, \n",
    "    save=False,\n",
    "    image_path=None,\n",
    "    meth_min_max_quantiles=(0.02, 0.98),\n",
    "    rna_min_max_quantiles=(0.02, 0.98),\n",
    "    vmin=0, vmax=1,\n",
    "    color_splits=None,\n",
    "    figsize=(8,4),\n",
    "    title_fontsize=14,\n",
    "    def_fontsize=12, \n",
    "):\n",
    "    \"\"\"Plot expression heatmap for given genes and cell type\"\"\"\n",
    "    import PyComplexHeatmap as pch\n",
    "    from matplotlib.colors import TwoSlopeNorm\n",
    "\n",
    "    if filename is None: \n",
    "        filename = f\"joint_expr_heatmap_{title.replace(' ', '_')}\"\n",
    "    if replicate_palette is None:\n",
    "        replicate_palette = adata_rna.uns.get(f\"{replicate_col}_palette\", None)\n",
    "\n",
    "    df_col = adata_rna.obs[['Group', 'MS_NORM', 'MS_compartment']].copy()\n",
    "    if df_col.shape[0] > 50000:\n",
    "        df_col = df_col.sample(50000, random_state=42)\n",
    "    if groupby_replicate:\n",
    "        df_col[replicate_col] = adata_rna.obs[replicate_col]\n",
    "        df_col = df_col.sort_values([replicate_col, 'MS_NORM'])\n",
    "    else:\n",
    "        df_col = df_col.sort_values('MS_NORM')\n",
    "\n",
    "    df_row = df_genes.iloc[:n_genes].copy()\n",
    "    if manual_added_genes is not None:\n",
    "        df_manual = df_genes[ df_genes['gene'].isin(manual_added_genes) ].copy()\n",
    "        df_row = pd.concat([df_row, df_manual], axis=0)\n",
    "    df_row.drop_duplicates(subset='gene', keep='first', inplace=True)\n",
    "    if \"gene\" in df_row.columns:\n",
    "        df_row = df_row.set_index('gene')\n",
    "    toplot = df_row.index[:gene_names]\n",
    "    if manual_added_genes is not None:\n",
    "        toplot = list(toplot) + manual_added_genes\n",
    "    df_row['annot'] = [c if c in toplot else np.nan for c in df_row.index]\n",
    "    df_row = df_row.sort_values(heatmap_order_col, ascending=True)\n",
    "    # if \"gene\" in df_row.columns:\n",
    "    #     df_row = df_row.set_index('gene')\n",
    "\n",
    "    if hasattr(adata_rna.X, \"toarray\"):\n",
    "        df_expr_rna = adata_rna.X.toarray()\n",
    "    else: \n",
    "        df_expr_rna = adata_rna.X\n",
    "    df_expr_rna = pd.DataFrame(df_expr_rna, index=adata_rna.obs_names, columns=adata_rna.var_names).T\n",
    "    df_expr_rna = df_expr_rna.loc[df_row.index, df_col.index]\n",
    "    # df_expr_norm_rna = (\n",
    "    #     df_expr_rna\n",
    "    #     .subtract(df_expr_rna.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    #     .div(df_expr_rna.quantile(min_max_quantiles[1], axis=1) - df_expr_rna.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    # )\n",
    "    repl = []\n",
    "    for cc, df in df_col.groupby(replicate_col, observed=True): \n",
    "        df_samp = df_expr_rna[ df.index ].copy()\n",
    "        df_norm = (\n",
    "            df_samp\n",
    "            .subtract(df_samp.quantile(rna_min_max_quantiles[0], axis=1), axis=0)\n",
    "            .div(df_samp.quantile(rna_min_max_quantiles[1], axis=1) - df_samp.quantile(rna_min_max_quantiles[0], axis=1), axis=0)\n",
    "        )\n",
    "        repl.append(df_norm)\n",
    "    df_expr_norm_rna =  pd.concat(repl, axis=1)\n",
    "\n",
    "    if hasattr(adata_mch.X, \"toarray\"):\n",
    "        df_expr_mch = adata_mch.X.toarray()\n",
    "    else: \n",
    "        df_expr_mch = adata_mch.X\n",
    "    df_expr_mch = pd.DataFrame(df_expr_mch, index=adata_mch.obs_names, columns=adata_mch.var_names).T\n",
    "    df_expr_mch = df_expr_mch.loc[df_row.index, df_col.index]\n",
    "    # df_expr_norm_mch = (\n",
    "    #     df_expr_mch\n",
    "    #     .subtract(df_expr_mch.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    #     .div(df_expr_mch.quantile(min_max_quantiles[1], axis=1) - df_expr_mch.quantile(min_max_quantiles[0], axis=1), axis=0)\n",
    "    # )\n",
    "    repl = []\n",
    "    for cc, df in df_col.groupby(replicate_col, observed=True): \n",
    "        df_samp = df_expr_mch[ df.index ].copy()\n",
    "        df_norm = (\n",
    "            df_samp\n",
    "            .subtract(df_samp.quantile(meth_min_max_quantiles[0], axis=1), axis=0)\n",
    "            .div(df_samp.quantile(meth_min_max_quantiles[1], axis=1) - df_samp.quantile(meth_min_max_quantiles[0], axis=1), axis=0)\n",
    "        )\n",
    "        repl.append(df_norm)\n",
    "    df_expr_norm_mch =  pd.concat(repl, axis=1)\n",
    "\n",
    "    ms_score_norm = TwoSlopeNorm(vmin=-1, vcenter=0, vmax=1)\n",
    "    col_ha = pch.HeatmapAnnotation(\n",
    "        Brain_Region=pch.anno_simple(df_col[replicate_col], colors=replicate_palette, add_text=True, legend=False, height=3, text_kws={\"fontsize\":8, \"fontweight\":\"bold\"}, label=\"Brain Region\"),\n",
    "        Compartment=pch.anno_simple(df_col['MS_compartment'], colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, height=3, label=\"Compartment\"),\n",
    "        MatStr_Score=pch.anno_simple(df_col['MS_NORM'], cmap=\"coolwarm_r\", norm=ms_score_norm, height=3, label=\"MatStr Score\"),\n",
    "        verbose=1, axis=1, plot_legend=True, legend_gap=5, label_kws={'visible':False}\n",
    "    )\n",
    "\n",
    "    left_ha = pch.HeatmapAnnotation(\n",
    "        label=pch.anno_label(\n",
    "            df_row['annot'], merge=True, rotation=0, extend=True,\n",
    "            colors=\"black\", relpos=(1, 0.7), height=3, fontsize=def_fontsize,\n",
    "            arrowprops = dict(visible=True, linewidth=0.5)\n",
    "        ),\n",
    "        verbose=1, axis=0\n",
    "    )\n",
    "\n",
    "    cm1 = pch.ClusterMapPlotter(\n",
    "        data=df_expr_norm_rna,\n",
    "        top_annotation=col_ha,\n",
    "        left_annotation=left_ha,\n",
    "        row_cluster=False,\n",
    "        col_cluster=False,\n",
    "        col_split=df_col[replicate_col] if groupby_replicate else None,\n",
    "        col_split_order=replicate_order,\n",
    "        row_dendrogram=False,\n",
    "        label=\"Expression\",\n",
    "        cmap=rna_cmap,\n",
    "        rasterized=rasterized,\n",
    "        ylabel=ylabel,\n",
    "        xlabel=\"RNA\",\n",
    "        xlabel_side=\"top\",\n",
    "        plot=False,\n",
    "        vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "\n",
    "    col_ha = pch.HeatmapAnnotation(\n",
    "        Brain_Region=pch.anno_simple(df_col[replicate_col], colors=replicate_palette, add_text=True, legend=False, height=3, text_kws={\"fontsize\":8, \"fontweight\":\"bold\"}, label=\"Brain Region\"),\n",
    "        Compartment=pch.anno_simple(df_col['MS_compartment'], colors={\"Matrix\": \"blue\", \"Striosome\": \"red\"}, height=3, label=\"Compartment\"),\n",
    "        MatStr_Score=pch.anno_simple(df_col['MS_NORM'], cmap=\"coolwarm_r\", norm=ms_score_norm, height=3, label=\"MatStr Score\"),\n",
    "        verbose=1, axis=1, plot_legend=True, legend_gap=5,\n",
    "    )\n",
    "\n",
    "    cm2 = pch.ClusterMapPlotter(\n",
    "        data=df_expr_norm_mch,\n",
    "        top_annotation=col_ha,\n",
    "        # left_annotation=left_ha,\n",
    "        row_cluster=False,\n",
    "        col_cluster=False,\n",
    "        col_split=df_col[replicate_col] if groupby_replicate else None,\n",
    "        col_split_order=replicate_order,\n",
    "        row_dendrogram=False,\n",
    "        col_dendrogram=False,\n",
    "        label=\"Score\",\n",
    "        cmap=mch_cmap,\n",
    "        rasterized=rasterized,\n",
    "        ylabel=ylabel,\n",
    "        xlabel=\"mCH\",\n",
    "        xlabel_side=\"top\",\n",
    "        plot=False,\n",
    "        vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "\n",
    "\n",
    "    plt.figure(figsize=figsize)\n",
    "    ax, legend_axes = pch.composite(cmlist=[cm1, cm2], main=0, legend_hpad=3, col_gap=0.1)\n",
    "    for _leg in legend_axes: \n",
    "        _leg.set_rasterized(False)\n",
    "    ax.set_title(title, fontsize=title_fontsize, rasterized=False)\n",
    "\n",
    "    # plot custom spines # For coloring the region in between the col splits as something other than white! \n",
    "    if color_splits is not None: \n",
    "        for cm in [cm1, cm2]:\n",
    "            for i in range(cm.heatmap_axes.shape[0]):\n",
    "                for j in range(cm.heatmap_axes.shape[1]-1):\n",
    "                    # if i != j:\n",
    "                    #     continue\n",
    "                    ax = cm.heatmap_axes[i][j]\n",
    "                    for side in [\"right\"]:\n",
    "                        ax.spines[side].set_visible(True)\n",
    "                        ax.spines[side].set_color(color_splits)\n",
    "                        ax.spines[side].set_linewidth(2)\n",
    "\n",
    "    if save and image_path is not None: \n",
    "        print(\"saving\")\n",
    "        plt.savefig(image_path / f\"{filename}.png\", dpi=300, bbox_inches=\"tight\")\n",
    "        plt.savefig(image_path / f\"{filename}.pdf\", dpi=300, bbox_inches=\"tight\", transparent=False)\n",
    "        # plt.savefig(image_path / f\"{filename}.svg\", dpi=300, bbox_inches=\"tight\")\n",
    "        # pass\n",
    "    if show: \n",
    "        plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d43c4cbf",
   "metadata": {},
   "source": [
    "# Spatial Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cf7e2ad",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13de3fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: @aklein Move to spida.pl\n",
    "def plot_categorical(\n",
    "    adata : ad.AnnData | str, \n",
    "    ax=None,\n",
    "    coord_base='tsne',\n",
    "    cluster_col='MajorType',\n",
    "    palette_path=None,\n",
    "    highlight_group : str | list | None = None,\n",
    "    color_unknown : str = \"#D0D0D0\",\n",
    "    is_background : bool = False,\n",
    "    axis_format='tiny',\n",
    "    alpha=0.7,\n",
    "    coding=True,\n",
    "    id_marker=True,\n",
    "    output=None,\n",
    "    show=True,\n",
    "    figsize=(4, 3.5),\n",
    "    sheet_name=None,\n",
    "    ncol=None,\n",
    "    fontsize=5,\n",
    "    legend_fontsize=5,\n",
    "    legend_kws=None,\n",
    "    legend_title_fontsize=5,\n",
    "    marker_fontsize=4,\n",
    "    marker_pad=0.1,\n",
    "    linewidth=0.5,\n",
    "    ret=False,\n",
    "    text_anno:bool = False,\n",
    "    text_kws=None,\n",
    "    **kwargs\n",
    "):\n",
    "    \"\"\"\n",
    "    Plot cluster.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata_path :\n",
    "    ax :\n",
    "    coord_base :\n",
    "    cluster_col :\n",
    "    palette_path :\n",
    "    highlight_group : str | list\n",
    "        A group (or list of groups) in the annotation to highlight by setting all other groups to background color (light gray).\n",
    "    color_unknown : str\n",
    "        Color to use for \"unknown\" or NaN categories. Default is light gray (#D0D0D0).\n",
    "    coding :\n",
    "    output :\n",
    "    show :\n",
    "    figsize :\n",
    "    sheet_name :\n",
    "    ncol :\n",
    "    fontsize :\n",
    "    legend_fontsize : int\n",
    "        legend fontsize, default 5\n",
    "    legend_kws: dict\n",
    "        kwargs passed to ax.legend\n",
    "    legend_title_fontsize: int\n",
    "        legend title fontsize, default 5\n",
    "    marker_fontsize: int\n",
    "        Marker fontsize, default 3\n",
    "        if id_marker is True, and coding is True. legend marker will be a circle (or rectangle) with code\n",
    "    linewidth : float\n",
    "        Line width of the legend marker (circle or rectangle), default 0.5\n",
    "    text_anon: bool\n",
    "        Whether to add text annotation for each cluster, default False.\n",
    "\tret : bool \n",
    "\t\tWhether to return the plot and colors, default False.\n",
    "    kwargs : dict\n",
    "        set text_anno=None to plot clustering without text annotations,\n",
    "        coding=True to plot clustering without code annotations,\n",
    "        set show_legend=False to remove the legend\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    \"\"\"\n",
    "    if sheet_name is None: # for getting color scheme from excel file\n",
    "        sheet_name=cluster_col\n",
    "    if isinstance(adata,str): # getting adata\n",
    "        adata=ad.read_h5ad(adata,backed='r')\n",
    "    if not isinstance(adata.obs[cluster_col].dtype, pd.CategoricalDtype): # make sure cluster_col is categorical \n",
    "        adata.obs[cluster_col] = adata.obs[cluster_col].astype('category')\n",
    "    # get palette\n",
    "    if palette_path is not None:\n",
    "        if isinstance(palette_path,str):\n",
    "            colors=pd.read_excel(os.path.expanduser(palette_path),sheet_name=sheet_name,index_col=0).Hex.to_dict()\n",
    "            keys=list(colors.keys())\n",
    "            existed_vals=adata.obs[cluster_col].unique().tolist()\n",
    "            for k in existed_vals:\n",
    "                if k not in keys:\n",
    "                    colors[k]='gray'\n",
    "            for k in keys:\n",
    "                if k not in existed_vals:\n",
    "                    del colors[k]\n",
    "        else:\n",
    "            colors=palette_path\n",
    "        adata.uns[cluster_col + '_colors'] = [colors.get(k, 'grey') for k in adata.obs[cluster_col].cat.categories.tolist()]\n",
    "    else:\n",
    "        if f'{cluster_col}_colors' not in adata.uns:\n",
    "            colors = add_color_scheme(adata, cluster_col, palette_key=f\"{cluster_col}_colors\")\n",
    "        else:\n",
    "            colors={cluster:color for cluster,color in zip(adata.obs[cluster_col].cat.categories.tolist(),adata.uns[f'{cluster_col}_colors'])}\n",
    "    data = adata[adata.obs[cluster_col].notna()]\n",
    "    if highlight_group is not None:\n",
    "        if isinstance(highlight_group, str):\n",
    "            highlight_group = [highlight_group]\n",
    "        all_groups = adata.obs[cluster_col].cat.categories.tolist()\n",
    "        for group in all_groups:\n",
    "            if group not in highlight_group:\n",
    "                colors[group] = color_unknown\n",
    "                data.obs.loc[data.obs[cluster_col] == group, cluster_col] = np.nan\n",
    "        data = data[data.obs[cluster_col].notna(), :]\n",
    "        data.obs[cluster_col] = data.obs[cluster_col].cat.remove_unused_categories()\n",
    "    colors = {k: v for k, v in colors.items() if k in data.obs[cluster_col].cat.categories.tolist()}\n",
    "    if is_background:\n",
    "        bg_colors = {k: color_unknown for k in colors.keys()}\n",
    "        bg_colors['unknown'] = color_unknown\n",
    "        colors = bg_colors        \n",
    "\n",
    "    hue=cluster_col\n",
    "    text_anno = cluster_col if text_anno else None\n",
    "    text_kws = {} if text_kws is None else text_kws\n",
    "    text_kws.setdefault(\"fontsize\", fontsize)\n",
    "    kwargs.setdefault(\"hue\",hue)\n",
    "    kwargs.setdefault(\"text_anno\", text_anno)\n",
    "    kwargs.setdefault(\"text_kws\", text_kws)\n",
    "    kwargs.setdefault(\"luminance\", 0.65)\n",
    "    kwargs.setdefault(\"dodge_text\", False)\n",
    "    kwargs.setdefault(\"axis_format\", axis_format)\n",
    "    kwargs.setdefault(\"show_legend\", True)\n",
    "    kwargs.setdefault(\"marker_fontsize\", marker_fontsize)\n",
    "    kwargs.setdefault(\"marker_pad\", marker_pad)\n",
    "    kwargs.setdefault(\"linewidth\", linewidth)\n",
    "    kwargs.setdefault(\"alpha\", alpha)\n",
    "    kwargs[\"coding\"]=coding\n",
    "    kwargs[\"id_marker\"]=id_marker\n",
    "    legend_kws={} if legend_kws is None else legend_kws\n",
    "    default_lgd_kws=dict(\n",
    "        fontsize=legend_fontsize,\n",
    "        title=cluster_col,title_fontsize=legend_title_fontsize)\n",
    "    if ncol is not None:\n",
    "        default_lgd_kws['ncol']=ncol\n",
    "    for k in default_lgd_kws:\n",
    "        legend_kws.setdefault(k, default_lgd_kws[k])\n",
    "    kwargs.setdefault(\"dodge_kws\", {\n",
    "            \"arrowprops\": {\n",
    "                \"arrowstyle\": \"->\",\n",
    "                \"fc\": 'grey',\n",
    "                \"ec\": \"none\",\n",
    "                \"connectionstyle\": \"angle,angleA=-90,angleB=180,rad=5\",\n",
    "            },\n",
    "            'autoalign': 'xy'})\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=figsize, dpi=300)\n",
    "    p = categorical_scatter(\n",
    "        data=data, #adata[adata.obs[cluster_col].notna(),],\n",
    "        ax=ax,\n",
    "        coord_base=coord_base,\n",
    "        palette=colors,\n",
    "        legend_kws=legend_kws,\n",
    "        **kwargs)\n",
    "\n",
    "    if output is not None:\n",
    "        plt.savefig(os.path.expanduser(output),bbox_inches='tight',dpi=300)\n",
    "    if show:\n",
    "        plt.show()\n",
    "    \n",
    "    if ret:\n",
    "        return p, colors\n",
    "    else: \n",
    "        return None, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9a6f22b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set up scale bar function\n",
    "\n",
    "def make_scale_bar(ax, x_coords, y_coords, microns_per_pixel=0.106, x_pct=100, y_pct=100):\n",
    "    \"\"\"\n",
    "    Adds a scale bar to a matplotlib plot.\n",
    "\n",
    "    Parameters:\n",
    "    - ax: matplotlib Axes object\n",
    "    - x_coords: list or array of x coordinates\n",
    "    - y_coords: list or array of y coordinates\n",
    "    - microns_per_pixel: conversion factor from pixels to microns\n",
    "    - x_pct: percentile of x-coordinates to position the scale bar horizontally\n",
    "    - y_pct: percentile of y-coordinates to position the scale bar vertically\n",
    "    \n",
    "    Example usage:\n",
    "    - fig, ax = plt.subplots()\n",
    "    - ax.scatter(x, y)\n",
    "    - make_scale_bar(ax, x, y)\n",
    "\n",
    "    \"\"\"\n",
    "    # Calculate x-axis range\n",
    "    x_range = [min(x_coords), max(x_coords)]\n",
    "    x_length = x_range[1] - x_range[0]\n",
    "    x_length_um = x_length * microns_per_pixel\n",
    "\n",
    "    # Target scale length ~1/6 of the x-axis\n",
    "    target = x_length_um / 6\n",
    "\n",
    "    # Compute order of magnitude\n",
    "    order = 10 ** np.floor(np.log10(target))\n",
    "    mantissa = target / order\n",
    "\n",
    "    # Round mantissa to nearest 1, 2, or 5\n",
    "    if mantissa < 1.5:\n",
    "        nice_mantissa = 1\n",
    "    elif mantissa < 3.5:\n",
    "        nice_mantissa = 2\n",
    "    elif mantissa < 7.5:\n",
    "        nice_mantissa = 5\n",
    "    else:\n",
    "        nice_mantissa = 10\n",
    "\n",
    "    # Final scale length in pixels\n",
    "    scale_length_um = nice_mantissa * order\n",
    "    scale_length_px = scale_length_um / microns_per_pixel\n",
    "\n",
    "    # Format label\n",
    "    if scale_length_um >= 1000:\n",
    "        scale_label = f\"{scale_length_um / 1000:.1f} mm\"\n",
    "    else:\n",
    "        scale_label = f\"{scale_length_um:.0f} µm\"\n",
    "\n",
    "    # Set coordinates for the scale bar\n",
    "    x_start = np.percentile(x_coords, x_pct) - scale_length_px * 1.1\n",
    "    x_end = np.percentile(x_coords, x_pct) - scale_length_px * 0.1\n",
    "    y_pos = np.percentile(y_coords, y_pct) - scale_length_px * 0.1\n",
    "    \n",
    "    # Set up background for scale bar\n",
    "    scale_bg = mpatches.Rectangle(\n",
    "        (x_start - scale_length_px * 0.05, y_pos - scale_length_px * 0.05),\n",
    "        width=(x_end - x_start) + scale_length_px * 0.1,\n",
    "        height=scale_length_px * 0.4,\n",
    "        color='white', alpha=0.8,\n",
    "        zorder = 10\n",
    "    )\n",
    "    \n",
    "    # Add background\n",
    "    ax.add_patch(scale_bg)\n",
    "\n",
    "    # Draw scale bar\n",
    "    ax.plot([x_start, x_end], [y_pos + scale_length_px * 0.3, y_pos + scale_length_px * 0.3], color='black', linewidth=2, zorder=11)\n",
    "    ax.text((x_start + x_end) / 2, y_pos + scale_length_px * 0.3, scale_label,\n",
    "            color='black', ha='center', va='bottom', zorder=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b0d3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_dense_df(X, var_names, obs_names):\n",
    "    \"\"\"Return a dense pandas DataFrame cells×genes from AnnData.X.\"\"\"\n",
    "    if hasattr(X, \"toarray\"):\n",
    "        arr = X.toarray()\n",
    "    elif hasattr(X, \"A\"):\n",
    "        arr = X.A\n",
    "    else:\n",
    "        arr = np.asarray(X)\n",
    "    return pd.DataFrame(arr, index=obs_names, columns=var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2b5a59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gene_groups(\n",
    "    adata: ad.AnnData,\n",
    "    gdf_geoms: gpd.GeoDataFrame,\n",
    "    gene: str,\n",
    "    groups: list[str],\n",
    "    regions: list[str],\n",
    "    group_level: str = \"Group\",\n",
    "    labs: list[str] = None,\n",
    "    donors: list[str] = None,\n",
    "    layers = None,\n",
    "    layer_norm = False,\n",
    "    color_wm = False,\n",
    "    hue_norm = 0.8,\n",
    "    min_norm = 0, \n",
    "    max_norm = 90,\n",
    "    image_path: str = None,\n",
    "    image_name: str = None,\n",
    "    rasterized: bool = False,\n",
    "    save_fig: bool = False,\n",
    "    show: bool = True, \n",
    "    shade_regions = False,\n",
    "    shade_alpha=0.05,\n",
    "    title_fontsize=16,\n",
    "    title = None,\n",
    "    show_scalebar = True,\n",
    "): \n",
    "    if gene not in adata.var_names:\n",
    "        print(f\"Gene {gene} not found in adata.var_names\")\n",
    "        return\n",
    "    adata = adata[adata.obs['brain_region'].isin(regions)].copy()\n",
    "    if layers is not None: \n",
    "        adata.X = adata.layers[layers].copy()\n",
    "    if layer_norm:\n",
    "        normalize_adata(adata, log1p=True)\n",
    "    \n",
    "    if labs is None: \n",
    "        labs = adata.obs['replicate'].unique().tolist()\n",
    "    if donors is None:\n",
    "        donors = adata.obs['donor'].unique().tolist()\n",
    "\n",
    "    nrows = len(labs)\n",
    "    ncols = len(regions)\n",
    "    nplots = len(donors)\n",
    "    \n",
    "    for i, donor in enumerate(donors):\n",
    "        fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(ncols*3, nrows*3), dpi=300)\n",
    "        # adata_don = adata[adata.obs['donor'] == donor].copy()\n",
    "        # hue_min = adata_don[:, gene].X.min()\n",
    "        # hue_max = adata_don[:, gene].X.max() * 0.9\n",
    "        for l, lab in enumerate(labs):\n",
    "            adata_don = adata[(adata.obs['donor'] == donor) & (adata.obs['replicate'] == lab)].copy()\n",
    "            X = to_dense_df(adata_don.X, adata_don.var_names, adata_don.obs_names)\n",
    "            X = X[gene].values\n",
    "            hue_min = np.percentile(X, min_norm)\n",
    "            hue_max = np.percentile(X, max_norm)\n",
    "            # hue_min = adata_don[:, gene].X.min()\n",
    "            # hue_max = adata_don[:, gene].X.max() * hue_norm\n",
    "            for r, region in enumerate(regions):\n",
    "                ax = axes[l, r] if nrows > 1 and ncols > 1 else axes # [max(l, r)]\n",
    "                # plot background cells\n",
    "                adata_sub = adata[(adata.obs['donor'] == donor) & (adata.obs['brain_region'] == region) & (adata.obs['replicate'] == lab)].copy()\n",
    "                categorical_scatter(adata_sub, coord_base=\"spatial\", color='lightgrey', ax=ax, rasterized=rasterized, axis_format=None)\n",
    "                \n",
    "                # Plot Data\n",
    "                adata_plot = adata_sub[adata_sub.obs[group_level].isin(groups)].copy()\n",
    "                if adata_plot.shape[0] != 0:\n",
    "                    plot_continuous(adata_plot, coord_base=\"spatial\", color_by=gene, cmap=\"YlOrRd\", ax=ax, show=False, hue_norm=(hue_min, hue_max), rasterized=rasterized, axis_format=None) \n",
    "                ax.set_title(f\"{region}\", rasterized=False)\n",
    "\n",
    "                # plot regions (MS vs. STR)\n",
    "                sub_geoms = gdf_geoms.loc[\n",
    "                    (gdf_geoms['brain_region'] == region) & \n",
    "                    (gdf_geoms['donor'] == donor) & \n",
    "                    (gdf_geoms['lab'] == lab)\n",
    "                    # (gdf_geoms['type'].isin([\"Striosome\", \"Matrix\"]))\n",
    "                ].copy()\n",
    "                sub_geoms.plot(ax=ax, edgecolor='black', facecolor='none', linewidth=1, rasterized=rasterized)\n",
    "                if color_wm: \n",
    "                    print(\"coloring_wm\")\n",
    "                    ss = sub_geoms[sub_geoms['type'] == \"White_Matter\"]\n",
    "                    print(ss.shape)\n",
    "                    ss.plot(ax=ax, color=color_wm, edgecolor=\"none\", rasterized=rasterized).axis(\"off\");\n",
    "                if shade_regions: \n",
    "                    sub_geoms.plot(ax=ax, color=sub_geoms['type_color'], edgecolor='none', legend=True, alpha=shade_alpha, rasterized=rasterized).axis(\"off\");\n",
    "                # sub_geoms.plot(ax=ax, color=\"none\", edgecolor='black', legend=True).axis(\"off\");\n",
    "        \n",
    "        # TODO: CHANGE THE WAY TITLE PLOTS\n",
    "        if nrows == 1 and ncols == 1: \n",
    "            print(\"A\")\n",
    "            if title is None: \n",
    "                ax.set_title(f\"{region} - {donor} - {gene}\", y=0.96, fontsize=title_fontsize, rasterized=False)\n",
    "            else: \n",
    "                ax.set_title(title, y=0.96, fontsize=title_fontsize, rasterized=False)\n",
    "            plt.suptitle(\"\")\n",
    "        else:\n",
    "            print(\"B\")\n",
    "            plt.suptitle(f\"Gene: {gene} Donor: {donor}, Expression\", y=1.02, fontsize=title_fontsize, rasterized=False)\n",
    "        if show_scalebar: \n",
    "            make_scale_bar(ax, adata_sub.obs['CENTER_X'], adata_sub.obs['CENTER_Y'], microns_per_pixel=1, x_pct=0, y_pct=0)\n",
    "        if save_fig and image_path is not None:\n",
    "            if image_name is not None: \n",
    "                image_save_path_png = f\"{image_path}/{image_name}.png\"\n",
    "                image_save_path_pdf = f\"{image_path}/{image_name}.pdf\"\n",
    "                # image_save_path_svg = f\"{image_path}/{image_name}.svg\"\n",
    "            else: \n",
    "                image_save_path_png = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.png\"\n",
    "                image_save_path_pdf = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.pdf\"\n",
    "                # image_save_path_svg = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.svg\"\n",
    "            plt.savefig(image_save_path_png, bbox_inches='tight', dpi=300)\n",
    "            plt.savefig(image_save_path_pdf, bbox_inches='tight', dpi=300)\n",
    "            # plt.savefig(image_save_path_svg, bbox_inches='tight', dpi=300)\n",
    "        if show: \n",
    "            plt.show()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da50a9d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gene_groups_meth(\n",
    "    adata: ad.AnnData,\n",
    "    gdf_geoms: gpd.GeoDataFrame,\n",
    "    gene: str,\n",
    "    groups: list[str],\n",
    "    regions: list[str],\n",
    "    group_level: str = \"Group\",\n",
    "    labs: list[str] = None,\n",
    "    donors: list[str] = None,\n",
    "    layers = None,\n",
    "    layer_norm = False,\n",
    "    color_wm = False,\n",
    "    max_norm = 90,\n",
    "    min_norm = 10,\n",
    "    bg_adata: ad.AnnData = None,\n",
    "    cmap=\"YlOrRd\",\n",
    "    save_fig: bool = False,\n",
    "    image_path: str = None,\n",
    "    image_name: str = None,\n",
    "    rasterized: bool = False,\n",
    "    show: bool = True, \n",
    "    shade_regions = False,\n",
    "    shade_alpha=0.05,\n",
    "    title_fontsize=16,\n",
    "    title = None,\n",
    "    show_scalebar = True,\n",
    "): \n",
    "    \n",
    "    if gene not in adata.var_names:\n",
    "        print(f\"Gene {gene} not found in adata.var_names\")\n",
    "        return\n",
    "    adata = adata[adata.obs['brain_region'].isin(regions)].copy()\n",
    "    if layers is not None: \n",
    "        adata.X = adata.layers[layers].copy()\n",
    "    if layer_norm:\n",
    "        normalize_adata(adata, log1p=True)\n",
    "    \n",
    "    if labs is None: \n",
    "        labs = adata.obs['replicate'].unique().tolist()\n",
    "    if donors is None:\n",
    "        donors = adata.obs['donor'].unique().tolist()\n",
    "\n",
    "    nrows = len(labs)\n",
    "    ncols = len(regions)\n",
    "    nplots = len(donors)\n",
    "    \n",
    "    for i, donor in enumerate(donors):\n",
    "        fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(ncols*3, nrows*3), dpi=300)\n",
    "        for l, lab in enumerate(labs):\n",
    "            adata_don = adata[(adata.obs['donor'] == donor) & (adata.obs['replicate'] == lab)].copy()\n",
    "            hue_min = np.percentile(adata_don[:, gene].X, min_norm)\n",
    "            hue_max = np.percentile(adata_don[:, gene].X, max_norm)\n",
    "            for r, region in enumerate(regions):\n",
    "                ax = axes[l, r] if nrows > 1 and ncols > 1 else axes # [max(l, r)]\n",
    "                adata_sub = adata[(adata.obs['donor'] == donor) & (adata.obs['brain_region'] == region) & (adata.obs['replicate'] == lab)].copy()\n",
    "                \n",
    "                # Plot bg\n",
    "                if bg_adata is not None: \n",
    "                    bg_sub = bg_adata[(bg_adata.obs['donor'] == donor) & (bg_adata.obs['brain_region'] == region) & (bg_adata.obs['replicate'] == lab)].copy()\n",
    "                    categorical_scatter(bg_sub, coord_base=\"spatial\", color='lightgrey', ax=ax, axis_format=None, rasterized=rasterized)\n",
    "                else: \n",
    "                    categorical_scatter(adata_sub, coord_base=\"spatial\", color='lightgrey', ax=ax, axis_format=None, rasterized=rasterized)\n",
    "\n",
    "                # Plot Data\n",
    "                adata_plot = adata_sub[adata_sub.obs[group_level].isin(groups)].copy()\n",
    "                if adata_plot.shape[0] != 0:\n",
    "                    plot_continuous(adata_plot, coord_base=\"spatial\", color_by=gene, cmap=cmap, ax=ax, show=False, hue_norm=(hue_min, hue_max), \n",
    "                                    rasterized=rasterized, axis_format=None) \n",
    "                \n",
    "                # plot regions (MS vs. STR)\n",
    "                sub_geoms = gdf_geoms.loc[\n",
    "                    (gdf_geoms['brain_region'] == region) & \n",
    "                    (gdf_geoms['donor'] == donor) & \n",
    "                    (gdf_geoms['lab'] == lab)\n",
    "                    # (gdf_geoms['type'].isin([\"Striosome\", \"Matrix\"]))\n",
    "                ].copy()\n",
    "                sub_geoms.plot(ax=ax, edgecolor='black', facecolor='none', linewidth=1, rasterized=rasterized)\n",
    "                if color_wm: \n",
    "                    print(\"coloring_wm\")\n",
    "                    ss = sub_geoms[sub_geoms['type'] == \"White_Matter\"]\n",
    "                    print(ss.shape)\n",
    "                    ss.plot(ax=ax, color=color_wm, edgecolor=\"none\", rasterized=rasterized).axis(\"off\");\n",
    "                if shade_regions: \n",
    "                    sub_geoms.plot(ax=ax, color=sub_geoms['type_color'], edgecolor='none', legend=True, alpha=shade_alpha, rasterized=rasterized).axis(\"off\");\n",
    "                # sub_geoms.plot(ax=ax, color=\"none\", edgecolor='black', legend=True).axis(\"off\");\n",
    "\n",
    "        # TODO: CHANGE THE WAY TITLE PLOTS\n",
    "        if nrows == 1 and ncols == 1: \n",
    "            print(\"A\")\n",
    "            if title is None: \n",
    "                ax.set_title(f\"{region} - {donor} - {gene}\", y=0.96, fontsize=title_fontsize, rasterized=False)\n",
    "            else: \n",
    "                ax.set_title(title, y=0.96, fontsize=title_fontsize, rasterized=False)\n",
    "            plt.suptitle(\"\")\n",
    "        else:\n",
    "            print(\"B\")\n",
    "            plt.suptitle(f\"Gene: {gene} Donor: {donor}, Expression\", y=1.02, fontsize=title_fontsize, rasterized=False)\n",
    "        if show_scalebar: \n",
    "            make_scale_bar(ax, adata_sub.obs['CENTER_X'], adata_sub.obs['CENTER_Y'], microns_per_pixel=1, x_pct=0, y_pct=0)\n",
    "        if save_fig and image_path is not None:\n",
    "            if image_name is not None: \n",
    "                image_save_path_png = f\"{image_path}/{image_name}.png\"\n",
    "                image_save_path_pdf = f\"{image_path}/{image_name}.pdf\"\n",
    "                # image_save_path_svg = f\"{image_path}/{image_name}.svg\"\n",
    "            else: \n",
    "                image_save_path_png = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.png\"\n",
    "                image_save_path_pdf = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.pdf\"\n",
    "                # image_save_path_svg = f\"{image_path}/gene_{gene}_donor_{donor}_spatial.svg\"\n",
    "            plt.savefig(image_save_path_png, bbox_inches='tight', dpi=300)\n",
    "            plt.savefig(image_save_path_pdf, bbox_inches='tight', dpi=300)\n",
    "            # plt.savefig(image_save_path_svg, bbox_inches='tight', dpi=300)\n",
    "        if show: \n",
    "            plt.show()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56cafb96",
   "metadata": {},
   "source": [
    "## Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dec07a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load Geoms path\n",
    "geom_store_path = \"/home/x-aklein2/projects/aklein/BICAN/BG/data/regions/region_geometries_cps.parquet\"\n",
    "gdf_geoms = gpd.read_parquet(geom_store_path)\n",
    "gdf_geoms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c43343d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['counts'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ceb96f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RASTERIZED = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8318f765",
   "metadata": {},
   "source": [
    "### STR D1 MSN \n",
    "\n",
    "PDYN + BACH2 good for Striosome\n",
    "\n",
    "KIRREL3 + GRM1 good for Matrix\n",
    "\n",
    "There are many more these are just two that I chose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "267d1d89",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"PDYN\", \"BACH2\", \"KIRREL3\", \"GRM1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db1c8ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR D1 MSN\"]\n",
    "_donor = \"UCI5224\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"PU\"]\n",
    "_gene = \"BACH2\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes: \n",
    "    plot_gene_groups(\n",
    "        adata,\n",
    "        gdf_geoms,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Subclass\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers='volume_norm',\n",
    "        layer_norm=True,\n",
    "        min_norm=0,\n",
    "        max_norm=99,\n",
    "        title_fontsize=14,\n",
    "        shade_regions = False, \n",
    "        color_wm = \"white\",\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        rasterized=RASTERIZED,\n",
    "        show=False,\n",
    "        save_fig=True,\n",
    "        show_scalebar=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52e1caa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR D1 MSN\"]\n",
    "_donor = \"UCI5224\"\n",
    "_lab = \"salk\"\n",
    "_regions = [\"PU\"]\n",
    "_gene = \"BACH2\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes: \n",
    "    plot_gene_groups_meth(\n",
    "        spatial_mch,\n",
    "        gdf_geoms,\n",
    "        bg_adata = adata,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Subclass\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers=None,\n",
    "        layer_norm=False,\n",
    "        cmap=\"Blues_r\",    \n",
    "        min_norm=20,\n",
    "        max_norm=80,\n",
    "        color_wm = 'white',\n",
    "        shade_regions = False, \n",
    "        shade_alpha=0.05,\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        title_fontsize=14,\n",
    "        save_fig=True,\n",
    "        show=False,\n",
    "        rasterized=RASTERIZED,\n",
    "        show_scalebar=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e31a5116",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclass_plot = [\"STR D1 MSN\"]\n",
    "# _donor = \"UCI5224\"\n",
    "# _lab = \"ucsd\"\n",
    "# _regions = [\"PU\"]\n",
    "# _gene = \"KIRREL3\"\n",
    "\n",
    "# subc_title = \", \".join(subclass_plot)\n",
    "# reg_title = \", \".join(_regions)\n",
    "\n",
    "# subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "# reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "# plot_gene_groups(\n",
    "#     adata,\n",
    "#     gdf_geoms,\n",
    "#     groups=subclass_plot,\n",
    "#     group_level=\"Subclass\", \n",
    "#     regions=_regions,\n",
    "#     donors=[_donor],\n",
    "#     labs = [_lab],\n",
    "#     gene=_gene,\n",
    "#     layers='volume_norm',\n",
    "#     layer_norm=True,\n",
    "#     hue_norm=0.5,\n",
    "#     shade_regions = False, \n",
    "#     color_wm = \"white\",\n",
    "#     image_path=image_path,\n",
    "#     image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "#     title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "#     title_fontsize=18,\n",
    "#     save_fig=True,\n",
    "#     show=False,\n",
    "#     rasterized=True\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "092293df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclass_plot = [\"STR D1 MSN\"]\n",
    "# _donor = \"UCI5224\"\n",
    "# _lab = \"salk\"\n",
    "# _regions = [\"PU\"]\n",
    "# _gene = \"BACH2\"\n",
    "\n",
    "# subc_title = \", \".join(subclass_plot)\n",
    "# reg_title = \", \".join(_regions)\n",
    "\n",
    "# subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "# reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "# plot_gene_groups_meth(\n",
    "#     spatial_mch,\n",
    "#     gdf_geoms,\n",
    "#     bg_adata = adata,\n",
    "#     groups=subclass_plot,\n",
    "#     group_level=\"Subclass\", \n",
    "#     regions=_regions,\n",
    "#     donors=[_donor],\n",
    "#     labs = [_lab],\n",
    "#     gene=_gene,\n",
    "#     layers=None,\n",
    "#     layer_norm=False,\n",
    "#     cmap=\"Blues_r\",    \n",
    "#     min_norm=20,\n",
    "#     max_norm=80,\n",
    "#     color_wm = 'white',\n",
    "#     shade_regions = False, \n",
    "#     shade_alpha=0.05,\n",
    "#     image_path=image_path,\n",
    "#     image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "#     title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "#     title_fontsize=18,\n",
    "#     save_fig=True,\n",
    "#     show=False,\n",
    "#     rasterized=True\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce28f306",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclass_plot = [\"STR D1 MSN\"]\n",
    "# _donor = \"UCI5224\"\n",
    "# _lab = \"salk\"\n",
    "# _regions = [\"PU\"]\n",
    "# _gene = \"KIRREL3\"\n",
    "\n",
    "# subc_title = \", \".join(subclass_plot)\n",
    "# reg_title = \", \".join(_regions)\n",
    "\n",
    "# subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "# reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "# plot_gene_groups_meth(\n",
    "#     spatial_mch,\n",
    "#     gdf_geoms,\n",
    "#     bg_adata = adata,\n",
    "#     groups=subclass_plot,\n",
    "#     group_level=\"Subclass\", \n",
    "#     regions=_regions,\n",
    "#     donors=[_donor],\n",
    "#     labs = [_lab],\n",
    "#     gene=_gene,\n",
    "#     layers=None,\n",
    "#     layer_norm=False,\n",
    "#     cmap=\"Blues_r\",    \n",
    "#     min_norm=20,\n",
    "#     max_norm=80,\n",
    "#     color_wm = 'white',\n",
    "#     shade_regions = False, \n",
    "#     shade_alpha=0.05,\n",
    "#     image_path=image_path,\n",
    "#     image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "#     title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "#     title_fontsize=18,\n",
    "#     show=False,\n",
    "#     save_fig=True,\n",
    "#     rasterized=True\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c45c7466",
   "metadata": {},
   "source": [
    "### CN ST18 GABA + STR FS PTHLH-PVALB GABA\n",
    "\n",
    "PLXNA4 + ALK + RASGRF2 for Striosome \n",
    "- PLXNA4 is a marker for *VIP GABA* interneurons\n",
    "- ALK is a marker for *STR-BF TAC3-PLPP4-LHX8 GABA* and *CN LAMP5-CXCL14 GABA*\n",
    "- RASGRF2 is a marker for *STR FS PTHLH-PVALB GABA*\n",
    "\n",
    "GLP1R + GALNT17 for Matrix\n",
    "- GLP1R is a marker for *CN ST18 GABA neurons*\n",
    "- GALNT17 is not a marker!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72688e69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subclass_plot = [\"CN ST18 GABA\"]\n",
    "# donors = [\"UCI5224\", \"UCI2424\", \"UCI4723\", \"UWA7648\"]\n",
    "# labs = [\"salk\", \"ucsd\"]\n",
    "# _regions = [\"CAH\", \"CAB\", \"PU\", \"NAC\"]\n",
    "# _gene = \"PLXNA4\"\n",
    "\n",
    "# subc_title = \", \".join(subclass_plot)\n",
    "# reg_title = \", \".join(_regions)\n",
    "\n",
    "# subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "# reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "# plot_gene_groups(\n",
    "#     adata,\n",
    "#     gdf_geoms,\n",
    "#     groups=subclass_plot,\n",
    "#     group_level=\"Subclass\", \n",
    "#     regions=_regions,\n",
    "#     donors= donors,\n",
    "#     labs = labs,\n",
    "#     gene=_gene,\n",
    "#     layers='volume_norm',\n",
    "#     layer_norm=False,\n",
    "#     min_norm=1,\n",
    "#     max_norm=99,\n",
    "#     title_fontsize=18,\n",
    "#     shade_regions = False, \n",
    "#     color_wm = \"white\",\n",
    "#     image_path=image_path,\n",
    "#     image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "#     title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "#     rasterized=True,\n",
    "#     show=True,\n",
    "#     save_fig=False,\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335cd1e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"PLXNA4\", \"ALK\", \"RASGRF2\", \"GLP1R\", \"GALNT17\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f41390f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"CN ST18 GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes:\n",
    "\n",
    "    plot_gene_groups(\n",
    "        adata,\n",
    "        gdf_geoms,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Subclass\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers='volume_norm',\n",
    "        layer_norm=False,\n",
    "        # hue_norm=0.5,\n",
    "        min_norm=0,\n",
    "        max_norm=99,\n",
    "        title_fontsize=14,\n",
    "        shade_regions = False, \n",
    "        color_wm = \"white\",\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        rasterized=RASTERIZED,\n",
    "        show=False,\n",
    "        save_fig=True,\n",
    "        show_scalebar=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc2f280",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"CN ST18 GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes:\n",
    "    plot_gene_groups_meth(\n",
    "        spatial_mch,\n",
    "        gdf_geoms,\n",
    "        bg_adata = adata,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Subclass\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers=None,\n",
    "        layer_norm=False,\n",
    "        cmap=\"Blues_r\",    \n",
    "        min_norm=2,\n",
    "        max_norm=98,\n",
    "        color_wm = 'white',\n",
    "        shade_regions = False, \n",
    "        shade_alpha=0.05,\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        title_fontsize=14,\n",
    "        show=False,\n",
    "        save_fig=True,\n",
    "        rasterized=RASTERIZED,\n",
    "        show_scalebar=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c728f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR FS PTHLH-PVALB GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "_gene = \"EPHA4\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes: \n",
    "    plot_gene_groups(\n",
    "        adata,\n",
    "        gdf_geoms,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Group\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers='volume_norm',\n",
    "        layer_norm=False,\n",
    "        # hue_norm=0.5,\n",
    "        min_norm=0,\n",
    "        max_norm=99,\n",
    "        title_fontsize=12,\n",
    "        shade_regions = False, \n",
    "        color_wm = \"white\",\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        rasterized=RASTERIZED,\n",
    "        show=False,\n",
    "        save_fig=True,\n",
    "        show_scalebar=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a5c5b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR FS PTHLH-PVALB GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "_gene = \"EPHA4\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "for _gene in genes: \n",
    "    plot_gene_groups_meth(\n",
    "        spatial_mch,\n",
    "        gdf_geoms,\n",
    "        bg_adata = adata,\n",
    "        groups=subclass_plot,\n",
    "        group_level=\"Group\", \n",
    "        regions=_regions,\n",
    "        donors=[_donor],\n",
    "        labs = [_lab],\n",
    "        gene=_gene,\n",
    "        layers=None,\n",
    "        layer_norm=False,\n",
    "        cmap=\"Blues_r\",\n",
    "        min_norm=1,\n",
    "        max_norm=90,\n",
    "        color_wm = 'white',\n",
    "        shade_regions = False, \n",
    "        shade_alpha=0.05,\n",
    "        image_path=image_path,\n",
    "        image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "        title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "        title_fontsize=12,\n",
    "        show=False,\n",
    "        save_fig=True,\n",
    "        rasterized=RASTERIZED,\n",
    "        show_scalebar=True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402b54f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR FS PTHLH-PVALB GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "_gene = \"GLP1R\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "plot_gene_groups(\n",
    "    adata,\n",
    "    gdf_geoms,\n",
    "    groups=subclass_plot,\n",
    "    group_level=\"Group\", \n",
    "    regions=_regions,\n",
    "    donors=[_donor],\n",
    "    labs = [_lab],\n",
    "    gene=_gene,\n",
    "    layers='volume_norm',\n",
    "    layer_norm=False,\n",
    "    # hue_norm=0.5,\n",
    "    min_norm=1,\n",
    "    max_norm=99,\n",
    "    title_fontsize=18,\n",
    "    shade_regions = False, \n",
    "    color_wm = \"white\",\n",
    "    image_path=image_path,\n",
    "    image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_rna',\n",
    "    title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "    rasterized=True,\n",
    "    show=True,\n",
    "    save_fig=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d90fb92",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclass_plot = [\"STR FS PTHLH-PVALB GABA\"]\n",
    "_donor = \"UWA7648\"\n",
    "_lab = \"ucsd\"\n",
    "_regions = [\"CAH\"]\n",
    "_gene = \"GLP1R\"\n",
    "\n",
    "subc_title = \", \".join(subclass_plot)\n",
    "reg_title = \", \".join(_regions)\n",
    "\n",
    "subc_filename = \"_\".join([s.replace(\" \", \"_\") for s in subclass_plot])\n",
    "reg_filename = \"_\".join([r.replace(\" \", \"_\") for r in _regions])\n",
    "\n",
    "plot_gene_groups_meth(\n",
    "    spatial_mch,\n",
    "    gdf_geoms,\n",
    "    bg_adata = adata,\n",
    "    groups=subclass_plot,\n",
    "    group_level=\"Group\", \n",
    "    regions=_regions,\n",
    "    donors=[_donor],\n",
    "    labs = [_lab],\n",
    "    gene=_gene,\n",
    "    layers=None,\n",
    "    layer_norm=False,\n",
    "    cmap=\"Blues_r\",    \n",
    "    min_norm=1,\n",
    "    max_norm=90,\n",
    "    color_wm = 'white',\n",
    "    shade_regions = False, \n",
    "    shade_alpha=0.05,\n",
    "    image_path=image_path,\n",
    "    image_name=f'ms_spatial_{subc_filename}_{_gene}_{_donor}_{_lab}_{reg_filename}_mch',\n",
    "    title = f\"{subc_title}\\n{reg_title} - {_gene}\",\n",
    "    title_fontsize=18,\n",
    "    show=True,\n",
    "    save_fig=False,\n",
    "    rasterized=True\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "preprocessing",
   "language": "python",
   "name": "preprocessing"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
